<!DOCTYPE html><html lang="zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0,viewport-fit=cover"><title>单细胞 | LeiLeiBear</title><meta name="author" content="Lei Lei"><meta name="copyright" content="Lei Lei"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="ffffff"><meta name="description" content="单细胞   index 标记样本，barcode 标记细胞，UMI 标记转录本。i7 sample index （library barcode）是加到 Illumina 测序接头上的，保证多个测序文库可以在同一个 flow-cell 上或者同一个 lane 上进行混合测序（multiplexed）。它的作用就是在 CellRanger 的 mkfastq 功能中体现出来的，它自动识别样本 ind">
<meta property="og:type" content="article">
<meta property="og:title" content="单细胞">
<meta property="og:url" content="https://glking123.github.io/2024/06/02/single-cell/index.html">
<meta property="og:site_name" content="LeiLeiBear">
<meta property="og:description" content="单细胞   index 标记样本，barcode 标记细胞，UMI 标记转录本。i7 sample index （library barcode）是加到 Illumina 测序接头上的，保证多个测序文库可以在同一个 flow-cell 上或者同一个 lane 上进行混合测序（multiplexed）。它的作用就是在 CellRanger 的 mkfastq 功能中体现出来的，它自动识别样本 ind">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://glking123.github.io/img/shengwu.jpg">
<meta property="article:published_time" content="2024-06-02T08:10:48.000Z">
<meta property="article:modified_time" content="2024-06-02T08:50:27.589Z">
<meta property="article:author" content="Lei Lei">
<meta property="article:tag" content="生物信息">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://glking123.github.io/img/shengwu.jpg"><link rel="shortcut icon" href="/img/animal.png"><link rel="canonical" href="https://glking123.github.io/2024/06/02/single-cell/index.html"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css?v=4.13.0"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@6.5.1/css/all.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/node-snackbar@0.1.16/dist/snackbar.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/ui@5.0.33/dist/fancybox/fancybox.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = {
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: undefined,
  noticeOutdate: undefined,
  highlight: {"plugin":"highlight.js","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false},
  copy: {
    success: '复制成功',
    error: '复制错误',
    noSupport: '浏览器不支持'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  dateSuffix: {
    just: '刚刚',
    min: '分钟前',
    hour: '小时前',
    day: '天前',
    month: '个月前'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: {"chs_to_cht":"你已切换为繁体中文","cht_to_chs":"你已切换为简体中文","day_to_night":"你已切换为深色模式","night_to_day":"你已切换为浅色模式","bgLight":"#49b1f5","bgDark":"#1f1f1f","position":"bottom-left"},
  infinitegrid: {
    js: 'https://cdn.jsdelivr.net/npm/@egjs/infinitegrid@4.11.1/dist/infinitegrid.min.js',
    buttonText: '加载更多'
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isAnchor: false,
  percent: {
    toc: true,
    rightside: false,
  },
  autoDarkmode: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: '单细胞',
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2024-06-02 16:50:27'
}</script><script>(win=>{
      win.saveToLocal = {
        set: (key, value, ttl) => {
          if (ttl === 0) return
          const now = Date.now()
          const expiry = now + ttl * 86400000
          const item = {
            value,
            expiry
          }
          localStorage.setItem(key, JSON.stringify(item))
        },
      
        get: key => {
          const itemStr = localStorage.getItem(key)
      
          if (!itemStr) {
            return undefined
          }
          const item = JSON.parse(itemStr)
          const now = Date.now()
      
          if (now > item.expiry) {
            localStorage.removeItem(key)
            return undefined
          }
          return item.value
        }
      }
    
      win.getScript = (url, attr = {}) => new Promise((resolve, reject) => {
        const script = document.createElement('script')
        script.src = url
        script.async = true
        script.onerror = reject
        script.onload = script.onreadystatechange = function() {
          const loadState = this.readyState
          if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
          script.onload = script.onreadystatechange = null
          resolve()
        }

        Object.keys(attr).forEach(key => {
          script.setAttribute(key, attr[key])
        })

        document.head.appendChild(script)
      })
    
      win.getCSS = (url, id = false) => new Promise((resolve, reject) => {
        const link = document.createElement('link')
        link.rel = 'stylesheet'
        link.href = url
        if (id) link.id = id
        link.onerror = reject
        link.onload = link.onreadystatechange = function() {
          const loadState = this.readyState
          if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
          link.onload = link.onreadystatechange = null
          resolve()
        }
        document.head.appendChild(link)
      })
    
      win.activateDarkMode = () => {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = () => {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', 'ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
        if (t === 'dark') activateDarkMode()
        else if (t === 'light') activateLightMode()
      
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    
      const detectApple = () => {
        if(/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)){
          document.documentElement.classList.add('apple')
        }
      }
      detectApple()
    })(window)</script><link rel="stylesheet" href="/css/head_1.css"><meta name="generator" content="Hexo 7.2.0"></head><body><script>window.paceOptions = {
  restartOnPushState: false
}

document.addEventListener('pjax:send', () => {
  Pace.restart()
})
</script><link rel="stylesheet" href="/css/flash.css"/><script src="https://cdn.jsdelivr.net/npm/pace-js@1.2.4/pace.min.js"></script><div id="web_bg"></div><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img is-center"><img src="/img/touxiang.jpg" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="sidebar-site-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">5</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">2</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">2</div></a></div><hr class="custom-hr"/><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> 主页</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> 时间轴</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> 标签</span></a></div><div class="menus_item"><a class="site-page group" href="javascript:void(0);"><i class="fa-fw fa fa-heartbeat"></i><span> 热爱</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/Gallery/"><i class="fa-fw fas fa-images"></i><span> 照片</span></a></li><li><a class="site-page child" href="/movies/"><i class="fa-fw fas fa-video"></i><span> 电影</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> 关于</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> 友链</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url('/img/shengwu.jpg')"><nav id="nav"><span id="blog-info"><a href="/" title="LeiLeiBear"><img class="site-icon" src="/img/animal.png"/><span class="site-name">LeiLeiBear</span></a></span><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> 主页</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> 时间轴</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> 标签</span></a></div><div class="menus_item"><a class="site-page group" href="javascript:void(0);"><i class="fa-fw fa fa-heartbeat"></i><span> 热爱</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/Gallery/"><i class="fa-fw fas fa-images"></i><span> 照片</span></a></li><li><a class="site-page child" href="/movies/"><i class="fa-fw fas fa-video"></i><span> 电影</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> 关于</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> 友链</span></a></div></div><div id="toggle-menu"><a class="site-page" href="javascript:void(0);"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">单细胞</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">发表于</span><time class="post-meta-date-created" datetime="2024-06-02T08:10:48.000Z" title="发表于 2024-06-02 16:10:48">2024-06-02</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">更新于</span><time class="post-meta-date-updated" datetime="2024-06-02T08:50:27.589Z" title="更新于 2024-06-02 16:50:27">2024-06-02</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF/">生物信息</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title="单细胞"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">阅读量:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><h1 id="单细胞"><a href="#单细胞" class="headerlink" title="单细胞"></a>单细胞</h1><p><img src="/post/single-cell/RTD3bITP1o6EJwxmULKceGz0nzo.png"></p>
<p><img src="/post/single-cell/BqGlbgRIoodcGnxM9dVcLpyqnbf.png"></p>
<p><img src="/post/single-cell/HxCEbWanEoZuRvxDoq2c9xENn7P.png"></p>
<p>index 标记样本，barcode 标记细胞，UMI 标记转录本。<br>i7 sample index （library barcode）<br>是加到 Illumina 测序接头上的，保证多个测序文库可以在同一个 flow-cell 上或者同一个 lane 上进行混合测序（multiplexed）。它的作用就是在 CellRanger 的 mkfastq 功能中体现出来的，它自动识别样本 index 名称（例如：SA-GA-A1），将具有相同 4 种 oligo 的 fq 文件组合在一起，表示同一个样本。它保证了一个测序 lane 上可以容纳多个样本。一个 index set 有 4 个 oligos。</p>
<p><strong>Barcode</strong><br>是 10X 特有的，用来区分 GEMs，也就是对细胞做了一个标记。一般在拆分混样测序数据（demultiplexing）这个过程后进行操作，当然这也很符合原文的操作。</p>
<p><strong>UMI</strong><br>UMI 就是 Unique Molecular Identifier，由 4-10 个随机核苷酸组成，在 mRNA 反转录后，进入到文库中，每一个 mRNA 随机连上一个 UMI，根据 PCR 结果可以计数不同的 UMI，最终统计 mRNA 的数量。它的主要作用是，处理 PCR 扩增偏差，因为起始文库很小时需要的 PCR 扩增次数就越多，因为越容易引入扩增误差。</p>
<p><img src="/post/single-cell/ChnXbJjZZodWFlxHkiQcqyHYngf.png"></p>
<p>n 单细胞测序（Single-cell sequencing）是一种基因组学技术，用于研究（某个组织）单个细胞的基因表达和基因组组成。传统的基因组学技术通常对大量细胞进行测序，从而获得平均表达水平。然而，细胞之间存在巨大的异质性，单细胞测序的出现使得研究人员能够更深入地了解每个单个细胞的特性和变异性。</p>
<p>单细胞测序技术可以通过两种主要方法来获得单细胞的基因组信息：</p>
<ol>
<li>单细胞 RNA 测序（Single-cell RNA sequencing，scRNA-seq）：这种方法用于分析单个细胞的转录组，即细胞中活跃的基因表达情况。它可以揭示不同细胞类型之间的差异、基因表达的动态变化以及细胞在发育、疾病等过程中的功能和调控机制。</li>
<li>单细胞 DNA 测序（Single-cell DNA sequencing，scDNA-seq）：这种方法可以分析单个细胞的基因组组成，包括基因突变、染色体重排等。通过单细胞 DNA 测序，可以研究细胞突变的发生、细胞演化以及肿瘤细胞异质性等重要问题。</li>
</ol>
<p>单细胞测序技术的应用非常广泛，尤其在发育生物学、免疫学、神经科学和肿瘤学等领域中具有重要意义。它可以帮助研究人员鉴定不同细胞亚型、构建细胞发育和分化的轨迹图、发现新的细胞类型和亚群，以及揭示细胞在疾病发生和发展中的变化。</p>
<p>单细胞测序技术的发展对于我们理解生命的复杂性和疾病的发生机制具有重要意义，它为个性化医学和精准治疗提供了新的途径。随着技术的不断进步，单细胞测序将继续为科学研究和医学领域带来更深入的认识和新的突破。</p>
<h1 id="单细胞转录组和转录组分析的区别和选择"><a href="#单细胞转录组和转录组分析的区别和选择" class="headerlink" title="单细胞转录组和转录组分析的区别和选择"></a>单细胞转录组和转录组分析的区别和选择</h1><p>单细胞转录组（single-cell transcriptome）是指对单个细胞中的所有转录本进行测序和分析的过程。它提供了对每个单个细胞的基因表达情况的全面了解，揭示了细胞间的异质性和多样性。</p>
<p>转录组分析是指对整个组织、细胞群体或样品中的转录本进行测序和分析的过程。它通常通过传统的高通量转录组测序方法，如 RNA-seq，来获取总体上的基因表达信息。</p>
<p>区别：</p>
<ol>
<li>尺度：单细胞转录组分析关注的是单个细胞的基因表达，而传统转录组分析则关注的是组织、细胞群体或样品整体的基因表达。</li>
<li>精度和分辨率：单细胞转录组分析具有更高的分辨率和精确性，可以揭示细胞间的异质性和个体细胞的差异。传统转录组分析往往掩盖了个体细胞之间的差异。</li>
<li>数据处理：由于单细胞转录组数据的特殊性，需要采用特定的数据处理和分析方法，如单细胞聚类和细胞发育轨迹推断等。传统转录组数据的处理方法通常适用于整体数据集的分析。</li>
</ol>
<p>选择：<br>选择单细胞转录组分析还是传统转录组分析取决于研究的问题和目标：</p>
<ol>
<li><p>单细胞转录组分析适用于以下情况：</p>
<ul>
<li>研究细胞类型的多样性和异质性。</li>
<li>探索细胞发育和分化的轨迹。</li>
<li>研究个体细胞之间的差异和变异。</li>
<li>发现和鉴定新的细胞亚群或亚型。</li>
</ul>
</li>
<li><p>传统转录组分析适用于以下情况：</p>
<ul>
<li>研究组织或细胞群体整体的基因表达变化。</li>
<li>需要在大规模样本中检测差异表达基因。</li>
<li>进行群体统计分析和功能注释。</li>
<li>比较不同条件下的整体基因表达。</li>
</ul>
</li>
</ol>
<p>综上所述，单细胞转录组和转录组分析各有其适用的场景和优势。根据研究的具体问题和需要，选择适当的分析方法可以更好地回答研究问题并获得所需的信息。</p>
<h1 id="空间转录组和单细胞转录组区别和选择"><a href="#空间转录组和单细胞转录组区别和选择" class="headerlink" title="空间转录组和单细胞转录组区别和选择"></a>空间转录组和单细胞转录组区别和选择</h1><p><img src="/post/single-cell/Pv85bDpHWoAWtoxnUyFcjJYHn5f.png"></p>
<p>空间转录组（Spatial transcriptomics）和单细胞转录组（Single-cell transcriptomics）都是用于研究基因表达的技术，但它们有一些区别和适用性上的差异。</p>
<p>区别：</p>
<ol>
<li>尺度：空间转录组关注的是组织或细胞在空间上的基因表达模式，可以提供基因表达的空间分布信息。而单细胞转录组关注的是单个细胞的基因表达，揭示细胞间的异质性和个体细胞的差异。</li>
<li>分辨率：空间转录组通常具有较高的空间分辨率，可以在组织或细胞级别上分析基因表达。而单细胞转录组则具有更高的分辨率和精确性，可以揭示细胞间更细微的差异和细胞类型的多样性。</li>
<li>数据类型：空间转录组通常提供图像或空间坐标数据，将基因表达信息与组织结构相对应。而单细胞转录组则生成基因表达矩阵，包含每个细胞中的基因表达量。</li>
</ol>
<p>选择：</p>
<p>选择空间转录组或单细胞转录组取决于研究问题和目标：</p>
<ol>
<li><p>空间转录组适用于以下情况：</p>
<ul>
<li>研究基因表达在组织内的空间分布和空间异质性。</li>
<li>探索细胞类型在组织结构中的定位和相互作用。</li>
<li>分析基因表达与细胞或组织结构的关联性。</li>
<li>研究组织发育、器官形成和组织重塑等过程。</li>
</ul>
</li>
<li><p>单细胞转录组适用于以下情况：</p>
<ul>
<li>研究细胞类型的多样性和异质性。</li>
<li>揭示细胞发育和分化的轨迹。</li>
<li>研究个体细胞之间的差异和变异。</li>
<li>发现和鉴定新的细胞亚群或亚型。</li>
</ul>
</li>
</ol>
<p>需要注意的是，近年来出现了一些整合空间信息和单细胞信息的技术，可以同时获得基因表达的空间分布和单细胞分辨率的基因表达信息。这些方法的发展使得研究人员能够更全面地了解基因表达的空间特征和细胞间的异质性。</p>
<p>综上所述，根据研究的具体问题和需要，选择适当的转录组技术可以更好地解答问题并获得所需的信息</p>
<h1 id="python-版"><a href="#python-版" class="headerlink" title="python 版"></a>python 版</h1><p>当进行单细胞测序数据的处理和分析时，以下是一些常见的 Python 库和工具，以及它们在每个步骤中的应用：</p>
<ol>
<li><p>数据预处理和质控：</p>
<ul>
<li><code>pandas</code>：用于加载、处理和管理数据框的库。</li>
<li><code>numpy</code>：用于处理数值计算和数组操作的库。</li>
<li><code>Scanpy</code>：一个用于单细胞数据分析的 Python 库，提供了质量控制、归一化和批次效应去除等功能。</li>
</ul>
</li>
<li><p>数据分析和聚类：</p>
<ul>
<li><code>scanpy</code>：用于单细胞数据分析的高级库，提供了聚类、降维、可视化等功能。</li>
<li><code>Seurat</code>：一个流行的 R 包，但也有 Python 版本的实现（SeuratWrappers），用于单细胞数据的聚类和可视化。</li>
<li><code>umap-learn</code>：用于高维数据降维和可视化的库，支持 t-SNE 和 UMAP 等算法。</li>
</ul>
</li>
<li><p>基因差异表达和功能富集分析：</p>
<ul>
<li><code>DESeq2</code>：用于差异表达分析的库，可帮助鉴定不同细胞群之间的显著差异基因。</li>
<li><code>scikit-learn</code>：一个流行的机器学习库，可用于分类和预测分析。</li>
<li><code>clusterProfiler</code>：用于功能富集分析的库，可帮助理解不同细胞群之间的生物学差异。</li>
</ul>
</li>
<li><p>可视化和结果展示：</p>
<ul>
<li><code>matplotlib</code>：一个功能强大的绘图库，用于绘制各种图表和图形。</li>
<li><code>seaborn</code>：一个基于 matplotlib 的统计数据可视化库，提供了更美观和易于使用的界面。</li>
<li><code>plotly</code>：一个交互式可视化库，可创建交互式的图表和可视化结果。</li>
</ul>
</li>
</ol>
<p>需要注意的是，单细胞测序数据分析是一个复杂的过程，可能涉及到更多的库和工具，具体取决于分析的要求和数据的特点。在实际操作中，您可能需要结合不同的工具和库，根据具体需求编写自定义的数据处理和分析代码。此外，许多单细胞测序分析平台和工具也提供了详细的文档和教程，可帮助您更好地理解和应用这些工具。</p>
<h1 id="1、单细胞测序发展历程"><a href="#1、单细胞测序发展历程" class="headerlink" title="1、单细胞测序发展历程"></a><strong>1、单细胞测序发展历程</strong></h1><p>那么为什么要进行单细胞测序呢？为什么要在传统转录组测序的基础上进行单细胞测序呢？因为组织、器官虽然是一个整体，但是组成他们细胞却是异质性的，也就是有不同的细胞类群。如果单纯做传统转录组，我们得到的是所有细胞的平均贡献，也许有些细胞的特殊贡献被湮没了。而不同的细胞发挥着不同的功能。也许在有些疾病模型中，在疾病进程中，细胞会出现转化，迁移。单细胞测序能够更好的抓住这些细节问题。每个组织之中有不同的细胞类型</p>
<p>单细胞测序最初要解决的两个重要的问题是单个细胞的捕获（细胞通量）和文库的建立（微量核酸扩增）。目前这两个问题已经得到了很好的解决，而且成本也在不断下降。单细胞捕获平台用的最多的是 10X Genomics 和 BD，他们有各自的建库方法，后期测序都不成问题。</p>
<h1 id="2、目前单细胞测序类型"><a href="#2、目前单细胞测序类型" class="headerlink" title="2、目前单细胞测序类型"></a><strong>2、目前单细胞测序类型</strong></h1><ul>
<li>单细胞转录组测序：单个细胞转录组水平测序</li>
<li>单细胞核测序：解决无法捕获完整细胞的测序策略</li>
<li>单细胞空间转录组测序：从组织到单细胞</li>
<li>单细胞 ATAC 测序：单细胞表观基因组学水平染色质可及性</li>
<li>单细胞蛋白组测序：目前不是特别成熟</li>
</ul>
<h2 id="10X-单细胞测序文件"><a href="#10X-单细胞测序文件" class="headerlink" title="*10X 单细胞测序文件*"></a>*<em>10X <strong><strong>单细胞测序</strong></strong>文件</em>*</h2><p>第一个是 barcode 文件，一个是 gene 文件，一个是 matrix 文件。</p>
<p><img src="/post/single-cell/Cmw4bDOq6oxHOYxtaPlcKzBJnWg.png"></p>
<p><img src="/post/single-cell/WMbzbDaO1okrYSxpXZocA37ZnIJ.png"></p>
<h1 id="单细胞转录组数据质控（QC）及合并去除批次效应"><a href="#单细胞转录组数据质控（QC）及合并去除批次效应" class="headerlink" title="单细胞转录组数据质控（QC）及合并去除批次效应"></a>单细胞转录组数据质控（QC）及合并去除批次效应</h1><p>一般下游分析 QC 包含：</p>
<ul>
<li>细胞基因检出数，低质量细胞基因检出数通常较低，双细胞或者同时捕获多个细胞会有很高的基因数。所以要去除低质量的，和过高的细胞。</li>
<li>细胞检测出的分子数</li>
<li>线粒体基因比例，一般低质量细胞或者死细胞线粒体基因检出数很高。但是特殊情况特殊对待，有些细胞功能活跃，线粒体活跃，检出数自然也会很高。所以不能一刀切。</li>
<li>注意：</li>
</ul>
<ol>
<li>在分析单细胞测序的 fastq 文件时，每个样本可能有多个 fastq 文件（例如用于不同的测序通道），但您通常仅需要正向和反向读取，而不是索引文件。</li>
<li>软件 Cellranger 可以用来为每个细胞和基因生成 UMI 计数矩阵，但仍需要使用 Seurat 或 Scanpy 等工具进行进一步处理。</li>
<li>如果要比较两个单细胞样本，则最好使用假批量方法，但也可以使用 Seurat 或 Scanpy 中的内置数据集函数。</li>
</ol>
<h1 id="cellranger-单细胞"><a href="#cellranger-单细胞" class="headerlink" title="cellranger 单细胞"></a>cellranger 单细胞</h1><p>cellranger 单细胞分析流程主要分为：数据拆分（<strong>cellranger mkfastq</strong>、细胞定量** cellranger count<strong>、组合分析</strong> cellranger aggr<strong>、参数调整</strong> cellranger reanalyze**。还有一些用户可能会用到的功能：mat2csv、mkgtf、mkref、vdj、mkvdjref、testrun、upload、sitecheck。</p>
<p><strong>filtered_gene_bc_matrices：是重要的一个目录，下面又包含了 barcodes.tsv.gz、features.tsv.gz、matrix.mtx.gz，是下游 Seurat、Scater、Monocle 等分析的输入文件。</strong><br>web_summary.html：质控比对报告（一般认为外显子的比对率要在 60% 以上）。barcode 用来标记细胞，UMI 用来标记转录本；其次，barcodes 数量时要大于细胞数量的（以保证每个细胞都会有 barcode 来进行区分）。</p>
<p>CellRanger 定量 count 过程输入文件指定命名格式为<br>[sample name]<em>S1_L00[lane number]</em>[read type]_001.fastq.gz<br>read type 有 3 种，I1 为 sample index read， R1 为 barcode 和 UMI， R2 才是测序 read。<br>例如：<br>pbmc_1k_v2_S1_L001_I1_001.fastq.gz<br>pbmc_1k_v2_S1_L001_R1_001.fastq.gz<br>pbmc_1k_v2_S1_L001_R2_001.fastq.gz<br>在 SRA 或 ENA 下载的数据要自己重命名为 cellranger 输入命名格式，<br>比如，将原来的 SRR7692286_1.fastq.gz 改成 SRR7692286_S1_L001_I1_001.fastq.gz</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">cat  SRR_Acc_List-<span class="number">9245</span>-<span class="number">3.</span>txt | <span class="keyword">while</span> read i ;do (mv $&#123;i&#125;_1*.gz $&#123;i&#125;_S1_L001_I1_001.fastq.gz;mv $&#123;i&#125;_2*.gz $&#123;i&#125;_S1_L001_R1_001.fastq.gz;mv $&#123;i&#125;_3*.gz $&#123;i&#125;_S1_L001_R2_001.fastq.gz);done</span><br></pre></td></tr></table></figure>

<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line">cellranger count --<span class="built_in">id</span> sample --transcriptome /path/refada_gex/ --fastqs /dasda/dasdsa/ </span><br><span class="line">--sample <span class="number">1</span> --expect-cells <span class="number">10000</span> --jobmode qsub -pe &lt;queue_name&gt; &lt;num_slots&gt; -l mem_free=&lt;memory&gt; -l h_rt=&lt;time_limit&gt;</span><br><span class="line"></span><br><span class="line">-pe &lt;queue_name&gt; &lt;num_slots&gt;用于指定使用的队列和所需的处理器槽位数。-l mem_free=&lt;memory&gt;用于指定可用的内存量，-l h_rt=&lt;time_limit&gt;用于指定作业的最大运行时间。</span><br><span class="line"></span><br><span class="line">这是一个关于使用cellranger count命令的帮助文档。cellranger count是一个用于计算基因表达和特征条形码读数的工具，用于处理单个样本和GEM孔的数据。</span><br><span class="line">你可以使用不同的选项来配置cellranger count命令的行为。下面是一些常用选项的说明：</span><br><span class="line">--<span class="built_in">id</span> &lt;ID&gt;：指定一个唯一的运行ID和输出文件夹名称。</span><br><span class="line">--transcriptome &lt;PATH&gt;：指定包含10x兼容转录组参考文件的文件夹路径。</span><br><span class="line">--fastqs &lt;PATH&gt;：指定输入FASTQ数据的路径。</span><br><span class="line">--sample &lt;PREFIX&gt;：指定要选择的FASTQ文件的前缀。</span><br><span class="line">--libraries &lt;CSV&gt;：指定包含输入库数据源的CSV文件。</span><br><span class="line">--feature-ref &lt;CSV&gt;：指定特征参考CSV文件，声明特征条形码构建和相关条形码。</span><br><span class="line">--target-panel &lt;CSV&gt;：指定目标面板CSV文件，声明使用的目标面板（如果有）。默认分析将排除内含子映射的读数，这是针对定向测序的推荐模式。使用include-introns=true可以包含内含子映射的读数进行分析。</span><br><span class="line">--expect-cells &lt;NUM&gt;：指定预计恢复的细胞数，作为细胞召回算法的输入。</span><br><span class="line">--force-cells &lt;NUM&gt;：强制使用指定的细胞数，绕过细胞召回算法。</span><br><span class="line">--no-bam：设置--no-bam以不生成BAM文件，这将减少管道的计算时间和输出目录的大小。如果不确定，请不要使用此选项。BAM文件可用于故障排除和下游分析。</span><br><span class="line">--nosecondary：禁用二次分析，例如聚类。</span><br><span class="line">--chemistry &lt;CHEM&gt;：测序方法配置。默认情况下，会自动检测测序方法，这是推荐的模式。通常不需要指定化学方法。选项包括：<span class="string">&#x27;auto&#x27;</span>（自动检测）、<span class="string">&#x27;threeprime&#x27;</span>（单细胞<span class="number">3</span><span class="string">&#x27;）、&#x27;</span>fiveprime<span class="string">&#x27;（单细胞5&#x27;</span>）、<span class="string">&#x27;SC3Pv1&#x27;</span>或<span class="string">&#x27;SC3Pv2&#x27;</span>或<span class="string">&#x27;SC3Pv3&#x27;</span>（单细胞<span class="number">3</span><span class="string">&#x27; v1/v2/v3）、&#x27;</span>SC3Pv3LT<span class="string">&#x27;（单细胞3&#x27;</span> v3 LT）、<span class="string">&#x27;SC3Pv3HT&#x27;</span>（单细胞<span class="number">3</span><span class="string">&#x27; v3 HT）、&#x27;</span>SC5P-PE<span class="string">&#x27;或&#x27;</span>SC5P-R2<span class="string">&#x27;（单细胞5&#x27;</span>，成对末端/R2-only）、<span class="string">&#x27;SC-FB&#x27;</span>（单细胞抗体-only <span class="number">3</span><span class="string">&#x27; v2或5&#x27;</span>）。要分析多组分数据的GEX部分，必须将化学方法设置为<span class="string">&#x27;ARC-v1&#x27;</span>；无法自动检测<span class="string">&#x27;ARC-v1&#x27;</span>化学方法。</span><br><span class="line">...（更多选项）</span><br><span class="line">--mempercore &lt;NUM&gt;：保留足够的线程给每个作业，以确保每个核心至少有这么多内存可用。仅适用于集群作业模式。</span><br><span class="line">--maxjobs &lt;NUM&gt;：设置一次提交给集群的最大作业数。仅适用于集群作业模式。</span><br><span class="line">--jobinterval &lt;NUM&gt;：设置在提交作业之间的延迟时间（以毫秒为单位）。仅适用于集群作业模式。</span><br><span class="line">你可以根据你的具体需求使用这些选项来运行cellranger count命令，进行基因表达和特征条形码读数的计算。</span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/RwCfbAck2oYrZlxflrhcHtE8nPf.png"></p>
<p><img src="/post/single-cell/B1WobPMTvoqDq1xVBaHc4pn5nJf.png"></p>
<h2 id="结果分析"><a href="#结果分析" class="headerlink" title="结果分析"></a>结果分析</h2><p><img src="/post/single-cell/SqG8bAWjCoEixsxSViAcjfNKnhd.png"></p>
<p>Estimated number of cells - 样本测到的细胞数量</p>
<p>Mean reads per cell - 每个细胞测到的平均 reads  （至少大于 20000）较低的指标是较低的测序深度</p>
<p>Median genes per cell - 每个细胞基因数的中位数</p>
<p><strong>sequencing</strong></p>
<p>Number of reads - 测到的总 read 数目</p>
<p>Valid barcodes - UMI 校正后匹配的 UMI 数量 **（大于 75%。较低的有效条形码比例可能表示存在测序问题（如低 Read 1 Q30 得分）。) **</p>
<p>*<em>Sequencing <strong><strong>saturation</strong></strong>：测序饱和度，一般 60-80% 比较合适（阈值范围可以适当调整，但是高于 70%&#x2F;80% 左右绝对 OK）较低的有效 UMI 比例可能表示存在测序或文库质量问题。</em>*</p>
<p>Q30 bases in barcode - 基于 barcode 的分数，大于 30 的比率</p>
<p>Q30 bases in RNA read - 基于 RNA read 的分数，大于 30 的比率</p>
<p>Q30 bases in UMI - 基于 UMI 的分数，大于 30 的比率</p>
<p>Median genes per cell&gt; 1000, Valid barcode &gt; 90%, Q30 &gt; 85%, Fraction reads &gt; 70%</p>
<p><img src="/post/single-cell/YTlwbZQM6oBB8CxGi4QcuzYTns4.png"></p>
<p>Reads mapped to genome - 比对到选定基因组的 reads</p>
<p>Reands mapped confidently to genome - 仅仅比对到基因组的 reads，如果一条 reads 既可以比对到外显子区又可以比对到非外显子区，那么算比对到了其中一个外显子区 大于 85%</p>
<p>Reads mapped confidently to intergenic regions - 比对到基因组的基因间区域</p>
<p>Reads mapped confidently to intronic regions - 比对到内含子区域</p>
<p>Reads mapped confidently to exonic regions - 比对到外显子区域</p>
<p>Reads mapped confidently to transcriptome - 比对到转录组的 reads，这些读数可以用来 UMI 的计数 预期值：变量，<strong>理想情况下大于 30%。</strong>参考质量和测序配置（Read 2 较短）会影响比对结果。较低的预期值可能表明使用了错误的参考转录组。</p>
<p>Reads mapped antisense to gene - 比对到基因的相反的 reads **理想情况下 &lt;10%**。如果使用预 mRNA 参考或使用了错误的 Gel Bead 化学反应，这些值可能较高。</p>
<p><img src="/post/single-cell/QOSPbzczxoLFvvxkrTachAEInec.png"></p>
<p><strong>要有急剧的下降说明分离的好</strong></p>
<p>比较重要的就是这个曲线图，图中横轴是 barcodes，纵轴是 UMI 数量，绿色线条代表真实的检测到的真实细胞数。通过 barcode 上的 UMI 标签分布来评估细胞数目，深蓝色代表细胞，灰色代表背景，示意如下。在前期磁珠（bead）与细胞形成油包水的结构过程中，会存在没有把细胞包进去的情况，这时候的油包水结构里面就只有磁珠和一些 barcode 的序列，而 cDNA 的碱基序列一般都是 barcode 碱基序列的 10 倍以上，就是由此来确定哪些是真实的细胞，哪些是 background。</p>
<p>Estimated number of cells - 样本测到的细胞数</p>
<p>Fraction reads in cells - valid-UMI 的质量分数，代表与细胞相关的 UMI 可靠地比对到基因组，一般要在 70% 及以上，否则数据质量就不好</p>
<p>Mean reads per cell - 每个细胞测到的平均 reads</p>
<p>Median genes per cell - 每个细胞的中间基因数</p>
<p>Total genes detected - 测到的总基因数，至少有一条 UMI</p>
<p>Median UMI counts per cell - 细胞 UMI 数量的中间值</p>
<p><img src="/post/single-cell/SJcdbm0fSoza65xnlwPcg7k1nue.png"></p>
<p><strong>该图显示了测序饱和度指标与<strong><strong>降采样</strong></strong>测序深度（以每个细胞平均读取数表示）之间的关系，直到观测到的测序深度为止。测序饱和度是对观测到的文库复杂性的衡量，当所有转录的 <strong><strong>mRNA</strong></strong> 已被测序时，该指标趋近于 1.0（100%）。曲线在接近端点处的<strong><strong>斜率</strong></strong>可以解释为从该点开始进一步增加测序深度所能获得的上限收益。虚线绘制在一个合理逼近饱和点的值处。</strong></p>
<p><img src="/post/single-cell/Quypb2Qdgopzldxh791cWYJ5nZb.png"></p>
<p>该图显示了每个细胞的中位基因数与平均每个细胞的下采样测序深度之间的关系，直到观察到的测序深度。曲线接近终点处的斜率可以解释为增加测序深度超过此点所能获得的收益的上限。</p>
<p>R-seurat</p>
<p><img src="/post/single-cell/XpWobdHLjoMx1UxlHLlce07knkh.png"></p>
<p>python-scanpy</p>
<p><img src="/post/single-cell/MkgNbAUzSohSYnxcTcicROPQnk5.png"></p>
<p><img src="/post/single-cell/RfjrbgTTqodtFOxFZKIczL2Hnlh.png"></p>
<p>表达的基因数目，总共测到多少表达量，线粒体基因表达的占比，细胞标签</p>
<p><img src="/post/single-cell/T1Nybvxg4oelizx4YCAc1W5gnzc.png"></p>
<p>基因在多少个细胞中表达了，是不是 mt 基因，</p>
<p><img src="/post/single-cell/E0vybnj3qoH4N4xaNb9cH12znue.png"></p>
<p><img src="/post/single-cell/USTEb8htbodM2VxiYdVcE6qPnxe.png"></p>
<p><img src="/post/single-cell/WscqbEX47omuKoxsiTUcPQ2SnZd.png"></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">cell_names = adata.obs.index.to_list()  <span class="comment"># 获取细胞名</span></span><br><span class="line">gene_names = adata.var.index.to_list()  <span class="comment"># 获取基因名</span></span><br><span class="line">num_cells = adata.n_obs  <span class="comment"># 细胞数目</span></span><br><span class="line">num_genes = adata.n_vars  <span class="comment"># 基因数目</span></span><br><span class="line"></span><br><span class="line">cluster_values = adata.obs[<span class="string">&#x27;cluster&#x27;</span>].values.tolist()[<span class="number">1</span>:<span class="number">5</span>]  <span class="comment"># 查看obs某列的值，若没有&#x27;cluster&#x27;列名则无法查看</span></span><br><span class="line"></span><br><span class="line">barcode_celltype_mapping = pd.DataFrame(&#123;<span class="string">&#x27;barcode&#x27;</span>: adata.obs.index.tolist(), <span class="string">&#x27;celltype&#x27;</span>: adata.obs.leiden.tolist()&#125;)  <span class="comment"># 可查看每个细胞barcode对应的细胞类型</span></span><br><span class="line"></span><br><span class="line">celltype_categories = adata.obs.leiden.values.categories.tolist()  <span class="comment"># 获取当前对象的细胞类型种类</span></span><br><span class="line"></span><br><span class="line">highly_variable_genes = adata[:, adata.var.highly_variable].var.index.tolist()  <span class="comment"># 获取高变异基因的索引</span></span><br><span class="line"></span><br><span class="line">adata.obs[<span class="string">&#x27;cluster&#x27;</span>] = adata.obs.leiden  <span class="comment"># 存储旧的细胞标签</span></span><br></pre></td></tr></table></figure>

<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"></span><br><span class="line"><span class="comment">#将adata.obs.leiden.values.tolist()的细胞聚类结果存储在DataFrame df 的 &#x27;cluster&#x27; 列中。</span></span><br><span class="line">df = pd.DataFrame(&#123;<span class="string">&#x27;cluster&#x27;</span>: adata.obs.leiden.values.tolist()&#125;)</span><br><span class="line"><span class="comment">#将聚类结果中的 &#x27;0&#x27; 聚类标签修改为 &#x27;CD4 T&#x27;，并将修改后的结果存储回 &#x27;cluster&#x27; 列</span></span><br><span class="line">df.loc[df.iloc[:, <span class="number">0</span>] == <span class="string">&#x27;0&#x27;</span>, <span class="number">0</span>] = <span class="string">&#x27;CD4 T&#x27;</span></span><br><span class="line"><span class="comment">#将修正后的聚类结果存储在 AnnData 对象的 &#x27;celltype&#x27; 列中。</span></span><br><span class="line">adata.obs[<span class="string">&#x27;celltype&#x27;</span>] = df.iloc[:, <span class="number">0</span>].tolist()</span><br><span class="line"><span class="comment">#将 &#x27;celltype&#x27; 列的数据类型转换为分类类型。</span></span><br><span class="line">adata.obs[<span class="string">&#x27;celltype&#x27;</span>] = adata.obs[<span class="string">&#x27;celltype&#x27;</span>].astype(<span class="string">&#x27;category&#x27;</span>)</span><br><span class="line"><span class="comment">#设置 &#x27;celltype&#x27; 列的分类顺序。</span></span><br><span class="line">adata.obs.celltype.cat.set_categories([<span class="string">&#x27;CD4 T&#x27;</span>, <span class="string">&#x27;2&#x27;</span>, <span class="string">&#x27;3&#x27;</span>, <span class="string">&#x27;4&#x27;</span>, <span class="string">&#x27;5&#x27;</span>, <span class="string">&#x27;6&#x27;</span>, <span class="string">&#x27;7&#x27;</span>, <span class="string">&#x27;8&#x27;</span>], inplace=<span class="literal">True</span>)</span><br><span class="line"><span class="comment">#根据 &#x27;cluster&#x27; 列的值筛选出对应的细胞。</span></span><br><span class="line">cell = adata.obs[adata.obs[<span class="string">&#x27;cluster&#x27;</span>] == <span class="string">&#x27;0&#x27;</span>].index</span><br><span class="line"><span class="comment">#根据 &#x27;cluster&#x27; 列的多个值筛选出对应的细胞。</span></span><br><span class="line">cell = adata.obs[adata.obs[<span class="string">&#x27;cluster&#x27;</span>].isin([<span class="string">&#x27;1&#x27;</span>, <span class="string">&#x27;2&#x27;</span>, <span class="string">&#x27;3&#x27;</span>, <span class="string">&#x27;4&#x27;</span>, <span class="string">&#x27;5&#x27;</span>, <span class="string">&#x27;6&#x27;</span>, <span class="string">&#x27;7&#x27;</span>, <span class="string">&#x27;8&#x27;</span>, <span class="string">&#x27;CD4 T&#x27;</span>])]</span><br></pre></td></tr></table></figure>

<h1 id="创建-anndata-对象-和-QC"><a href="#创建-anndata-对象-和-QC" class="headerlink" title="创建 anndata 对象 和 QC"></a>创建 anndata 对象 和 QC</h1><p>10X 数据读入和创建 anndata 对象：</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 载入包</span></span><br><span class="line"> <span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"> <span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"> <span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"> </span><br><span class="line"> <span class="comment"># 设置</span></span><br><span class="line"> sc.settings.verbosity = <span class="number">3</span>             </span><br><span class="line"><span class="comment"># 这行代码设置了日志的等级，控制在运行过程中显示的日志信息。日志等级有四个选项：</span></span><br><span class="line"><span class="comment">#0仅显示错误信息。</span></span><br><span class="line"><span class="comment">#1：显示警告和错误信息。</span></span><br><span class="line"><span class="comment">#2：显示信息、警告和错误信息。</span></span><br><span class="line"><span class="comment">#3：显示提示、信息、警告和错误信息。</span></span><br><span class="line"><span class="comment">#在这里，将日志等级设置为3，表示显示提示、信息、警告和错误信息。</span></span><br><span class="line"> sc.logging.print_header()</span><br><span class="line"><span class="comment">#这行代码用于打印Scanpy的日志标题。</span></span><br><span class="line"><span class="comment">#它会在控制台中显示一条信息，指示使用的Scanpy版本和相关信息。</span></span><br><span class="line"> sc.settings.set_figure_params(dpi=<span class="number">80</span>, facecolor=<span class="string">&#x27;white&#x27;</span>)</span><br><span class="line"><span class="comment">#这行代码用于设置绘图参数，以控制生成的图形的质量和外观。</span></span><br><span class="line"><span class="comment">#dpi=80设置绘图的分辨率为80个像素每英寸，可以根据需要进行调整。</span></span><br><span class="line"><span class="comment">#facecolor=&#x27;white&#x27;设置绘图的背景颜色为白色。</span></span><br><span class="line"> </span><br><span class="line"> <span class="comment"># 用于存储分析结果文件的路径</span></span><br><span class="line"> results_file = <span class="string">&#x27;write/pbmc3k.h5ad&#x27;</span></span><br><span class="line"> </span><br><span class="line"> <span class="comment"># 载入文件</span></span><br><span class="line"> adata = sc.read_10x_mtx(</span><br><span class="line">     <span class="string">&#x27;./filtered_gene_bc_matrices/hg19/&#x27;</span>,  <span class="comment"># mtx 文件目录</span></span><br><span class="line">     var_names=<span class="string">&#x27;gene_symbols&#x27;</span>,             <span class="comment"># 使用 gene_symbols 作为变量名</span></span><br><span class="line">     cache=<span class="literal">True</span>)                           <span class="comment"># 写入缓存，可以更快的读取文件</span></span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/SKGfbE3vUoJqZ0xkNd3coZVLnVg.png"></p>
<p><img src="/post/single-cell/V3IabZPWZoLGw0xPuR0crOFSnjd.png"></p>
<p><img src="/post/single-cell/TPPVb6LQoo6lHzxMkF9cd7dOnfh.png"></p>
<h2 id="预处理"><a href="#预处理" class="headerlink" title="**预处理 **"></a>**预处理 **</h2><h3 id="显示在所有细胞中在每个单细胞中产生最高计数分数的基因"><a href="#显示在所有细胞中在每个单细胞中产生最高计数分数的基因" class="headerlink" title="显示在所有细胞中在每个单细胞中产生最高计数分数的基因"></a><strong>显示在所有细胞中在每个单细胞中产生最高计数分数的基因</strong></h3><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line">adata.var_names_make_unique()</span><br><span class="line"> adata.obs_names_make_unique()</span><br><span class="line"> sc.pl.highest_expr_genes(adata, n_top=<span class="number">20</span>, )</span><br><span class="line"> <span class="comment">#您还需要通过设置n_top参数来指定您想要查看每个细胞或簇的前N个顶级基因数。</span></span><br><span class="line"> <span class="comment">#可视化每个细胞或簇中表达最高的前20个基因</span></span><br><span class="line"> <span class="comment">#这将生成一个图表，显示数据集中平均表达水平最高的前20个基因。</span></span><br><span class="line"> <span class="comment">#figsize（元组）：定图形的尺寸，以英寸为单位。例如，figsize=(6, 4) 表示图形的宽度为6英寸，高度为4英寸。</span></span><br><span class="line"><span class="comment">#show（布尔值）：指定是否显示图形。默认值为 True，将图形显示在屏幕上。如果将其设置为 False，图形将不会显示，只会返回图形对象。</span></span><br><span class="line"><span class="comment">#save（布尔值或字符串）：指定是否将图形保存到文件。如果设置为 True，图形将根据文件名规则保存到默认位置。如果提供一个字符串作为文件路径，图形将保存到指定的文件位置。</span></span><br><span class="line"><span class="comment">#title（字符串）：设置图形的标题。</span></span><br><span class="line"><span class="comment">#xlabel（字符串）：设置x轴的标签。</span></span><br><span class="line"><span class="comment">#ylabel（字符串）：设置y轴的标签。</span></span><br><span class="line"><span class="comment">#color_map（字符串）：设置条形图的颜色映射。可以使用预定义的颜色映射名称，例如 &#x27;viridis&#x27;、&#x27;magma&#x27;、&#x27;coolwarm&#x27; 等。</span></span><br><span class="line">sc.pl.highest_expr_genes(</span><br><span class="line">    adata,</span><br><span class="line">    n_top=<span class="number">20</span>,</span><br><span class="line">    figsize=(<span class="number">6</span>, <span class="number">4</span>),</span><br><span class="line">    show=<span class="literal">False</span>,</span><br><span class="line">    save=<span class="string">&#x27;highest_expr_genes.png&#x27;</span>,</span><br><span class="line">    title=<span class="string">&#x27;Highest Expressed Genes&#x27;</span>,</span><br><span class="line">    xlabel=<span class="string">&#x27;Genes&#x27;</span>,</span><br><span class="line">    ylabel=<span class="string">&#x27;Expression Level&#x27;</span>,</span><br><span class="line">    color_map=<span class="string">&#x27;viridis&#x27;</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/Gcmhbz8q5olm5IxXFhDcCqRAnzb.png"></p>
<h3 id="过滤低质量细胞样本-pp-preprocessing"><a href="#过滤低质量细胞样本-pp-preprocessing" class="headerlink" title="过滤低质量细胞样本 pp preprocessing"></a><strong>过滤低质量细胞样本 <strong><strong>pp</strong></strong> preprocessing</strong></h3><p><strong>质量控制的第一步是过滤低质量的细胞。当细胞检测到的基因数量较少、计数深度较低且<strong><strong>线粒体</strong></strong>计数较高时，细胞膜可能会破裂，这表明细胞正在死亡。由于这些细胞通常不是我们分析的主要目标，并且可能会扭曲我们的下游分析，因此我们在质量控制过程中将其去除。</strong></p>
<p><strong>过滤在少于三个细胞中表达，或一个细胞中表达少于 200 个基因的细胞样本</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">sc.pp.filter_cells(adata, min_genes=<span class="number">200</span>)</span><br><span class="line"><span class="comment">#用于过滤细胞，只保留至少具有指定数量基因表达的细胞。具体参数如下：</span></span><br><span class="line"><span class="comment">#adata：AnnData 对象，包含单细胞数据。</span></span><br><span class="line"><span class="comment">#min_genes（整数，可选）：要保留的最低基因表达数。默认值为 200。</span></span><br><span class="line"></span><br><span class="line">sc.pp.filter_cells()：</span><br><span class="line"><span class="comment">#min_genes：要保留的最低基因表达数。</span></span><br><span class="line"><span class="comment">#max_genes：要保留的最高基因表达数。</span></span><br><span class="line"><span class="comment">#min_counts：要保留的最低计数数。</span></span><br><span class="line"><span class="comment">#max_counts：要保留的最高计数数。</span></span><br><span class="line"><span class="comment">#min_mito：要保留的最低线粒体基因比例。</span></span><br><span class="line"><span class="comment">#max_mito：要保留的最高线粒体基因比例。</span></span><br><span class="line"></span><br><span class="line"> sc.pp.filter_genes(adata, min_cells=<span class="number">3</span>)</span><br><span class="line"><span class="comment">#sc.pp.filter_genes(adata, min_cells=3) 用于过滤基因，只保留至少在指定数量细胞中表达的基因。具体参数如下：</span></span><br><span class="line"><span class="comment">#adata：AnnData 对象，包含单细胞数据。</span></span><br><span class="line"><span class="comment">#min_cells（整数，可选）：要保留的最低细胞数，即至少在多少个细胞中表达的基因才会被保留。默认值为 3</span></span><br><span class="line"></span><br><span class="line">sc.pp.filter_genes()：</span><br><span class="line"><span class="comment">#min_cells：要保留的最低细胞数，即至少在多少个细胞中表达的基因才会被保留。</span></span><br><span class="line"><span class="comment">#max_cells：要保留的最高细胞数，即在多于指定数量的细胞中表达的基因将被过滤掉。</span></span><br><span class="line"><span class="comment">#min_counts：要保留的最低计数数。</span></span><br><span class="line"><span class="comment">#max_counts：要保留的最高计数数。</span></span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/NRaGbsqv6ozfcHxAOrxc48TWn3f.png"></p>
<h3 id="过滤包含线粒体基因和表达基因过多的细胞"><a href="#过滤包含线粒体基因和表达基因过多的细胞" class="headerlink" title="过滤包含线粒体基因和表达基因过多的细胞"></a><strong>过滤包含<strong><strong>线粒体</strong></strong>基因和表达基因过多的细胞</strong></h3><p>*<em>如果一个细胞正在死亡，那么其</em><em><strong>mRNA</strong></em><em>被释放到内环境，导致</em><em><strong>线粒体</strong></em><em>基因的比例较高，所以我们可以通过线粒体基因的比例来过滤掉低质量的</em><em><strong>单细胞测序</strong></em><em>数据。但是如果仅考虑一个变量可能会造成生物学误差，共同考虑三个 <strong><strong>QC</strong></strong> <strong><strong>协变量</strong></strong>至关重要。</em>*</p>
<p>线粒体基因的转录本比单个转录物分子大，并且不太可能通过细胞膜逃逸。因此，检测出高比例的线粒体基因，表明细胞质量差。</p>
<p>表达基因过多可能是由于一个油滴包裹多个细胞，从而检测出比正常检测要多的基因数，因此要过滤这些细胞。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">adata.var[<span class="string">&#x27;mt&#x27;</span>] = adata.var_names.<span class="built_in">str</span>.startswith(<span class="string">&#x27;MT-&#x27;</span>)  <span class="comment"># 将线粒体基因标记为 mt</span></span><br><span class="line"> sc.pp.calculate_qc_metrics(adata, qc_vars=[<span class="string">&#x27;mt&#x27;</span>], percent_top=<span class="literal">None</span>, log1p=<span class="literal">True</span>, inplace=<span class="literal">True</span>)</span><br><span class="line"> <span class="comment">#qc_vars：指定要计算的质控指标。可以是一个包含变量名的列表，例如 [&#x27;mt&#x27;, &#x27;gene&#x27;, &#x27;other_var&#x27;]，也可以是一个字符串，例如 &#x27;mt&#x27;。默认为 None，表示计算所有可用的质控指标。</span></span><br><span class="line"> <span class="comment">#percent_top：计算基因表达量的前百分之几。例如，percent_top=10 表示计算基因表达量排名前 10% 的基因。默认为 None，表示计算所有基因。</span></span><br><span class="line"> <span class="comment">#log1p：是否对计算结果进行 log1p 转换。默认为 False，表示不进行转换。</span></span><br><span class="line"> <span class="comment">#inplace：是否将计算结果保存在原始数据中。默认为 True，表示在原始数据上进行修改。</span></span><br><span class="line"> sc.pl.violin(adata, [<span class="string">&#x27;n_genes_by_counts&#x27;</span>, <span class="string">&#x27;total_counts&#x27;</span>, <span class="string">&#x27;pct_counts_mt&#x27;</span>],jitter=<span class="number">0.4</span>, multi_panel=<span class="literal">True</span>,show=<span class="literal">False</span>,</span><br><span class="line">    save=<span class="string">&#x27;highest_expr_genes.png&#x27;</span>,)</span><br><span class="line"><span class="comment">#&#x27;n_genes_by_counts&#x27; 表示每个细胞的基因数目，</span></span><br><span class="line"><span class="comment">#&#x27;total_counts&#x27; 表示每个细胞的总计数，&#x27;pct_counts_mt&#x27; 表示每个细胞中线粒体基因的百分比</span></span><br><span class="line"><span class="comment">#groupby：按照指定的变量进行分组，每个分组绘制一个小提琴图。</span></span><br><span class="line"><span class="comment">#use_raw：是否使用原始数据进行绘图。默认为 False，表示使用经过预处理的数据进行绘图。</span></span><br><span class="line"><span class="comment">#stripplot：是否在小提琴图上添加散点图。默认为 False，表示不添加散点图。</span></span><br><span class="line"><span class="comment">#jitter：散点图的抖动大小。设置为 0 表示不抖动，设置为大于 0 的值表示抖动的大小。</span></span><br><span class="line"><span class="comment">#size：散点图点的大小。</span></span><br><span class="line"><span class="comment">#log：是否对数值进行对数转换。默认为 False。</span></span><br><span class="line"><span class="comment">#show：是否显示图形。默认为 True。</span></span><br><span class="line"><span class="comment">#save：是否保存图形。默认为 False。</span></span><br></pre></td></tr></table></figure>

<p>生成的三张小提琴图代表：表达基因的数量，每个细胞包含的表达量，线粒体基因表达量的百分比。</p>
<p><img src="/post/single-cell/M0JwbEadcownNcxYh4dcFvwOncd.png"></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> matplotlib.pyplot <span class="keyword">as</span> plt</span><br><span class="line">mito_filter = <span class="number">15</span></span><br><span class="line">n_counts_filter = <span class="number">4300</span></span><br><span class="line">fig, axs = plt.subplots(ncols = <span class="number">2</span>, figsize = (<span class="number">8</span>,<span class="number">4</span>))</span><br><span class="line">sc.pl.scatter(adata, x=<span class="string">&#x27;total_counts&#x27;</span>, y=<span class="string">&#x27;pct_counts_mt&#x27;</span>,ax = axs[<span class="number">0</span>], show=<span class="literal">False</span>)</span><br><span class="line">sc.pl.scatter(adata, x=<span class="string">&#x27;total_counts&#x27;</span>, y=<span class="string">&#x27;n_genes_by_counts&#x27;</span>,ax = axs[<span class="number">1</span>], show = <span class="literal">False</span>)</span><br><span class="line"><span class="comment">#draw horizontal red lines indicating thresholds.</span></span><br><span class="line">axs[<span class="number">0</span>].hlines(y = mito_filter, xmin = <span class="number">0</span>, xmax = <span class="built_in">max</span>(adata.obs[<span class="string">&#x27;total_counts&#x27;</span>]), color = <span class="string">&#x27;red&#x27;</span>, ls = <span class="string">&#x27;dashed&#x27;</span>)</span><br><span class="line">axs[<span class="number">1</span>].hlines(y = n_counts_filter, xmin = <span class="number">0</span>, xmax = <span class="built_in">max</span>(adata.obs[<span class="string">&#x27;total_counts&#x27;</span>]), color = <span class="string">&#x27;red&#x27;</span>, ls = <span class="string">&#x27;dashed&#x27;</span>)</span><br><span class="line">fig.tight_layout()</span><br><span class="line">plt.show()</span><br></pre></td></tr></table></figure>

<p>过滤：</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 获取线粒体基因占比在 5% 以下的细胞样本</span></span><br><span class="line"> adata = adata[adata.obs.pct_counts_mt &lt; <span class="number">5</span>, :]</span><br><span class="line"> <span class="comment"># 获取表达基因数在 2500 以下的细胞样本</span></span><br><span class="line"> adata = adata[adata.obs.n_genes_by_counts &lt; <span class="number">2500</span>, :]</span><br></pre></td></tr></table></figure>

<p>双细胞被定义为在相同的细胞条形码（barcode）下进行测序的两个细胞，例如，如果它们被捕获在同一个液滴（droplet）中。这也是为什么我们一直使用 barcode 而不是 cells 的原因。双细胞由同型（homotypic）与异型（heterotypic）所构成</p>
<ul>
<li>同型：同型通常被认为是不影响下游分析的，因为其是由一类相同的细胞中的两个所构成，所以这部分细胞不是我们所需要过滤的对象</li>
<li>异型：异型通常是由来自两类不同的细胞所构成的，异型的存在会使得我们后续的细胞分类出现错误，因为其独特的数据分布特征。</li>
</ul>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">n0 = adata.shape[<span class="number">0</span>]</span><br><span class="line"><span class="built_in">print</span>(<span class="string">f&#x27;Original cell number: <span class="subst">&#123;n0&#125;</span>&#x27;</span>)</span><br></pre></td></tr></table></figure>

<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">print</span>(<span class="string">&#x27;Begin of post doublets removal and QC plot&#x27;</span>)</span><br><span class="line">sc.external.pp.scrublet(adata, random_state=<span class="number">112</span>)</span><br><span class="line">adata = adata[adata.obs[<span class="string">&#x27;predicted_doublet&#x27;</span>]==<span class="literal">False</span>, :].copy()</span><br><span class="line">n1 = adata.shape[<span class="number">0</span>]</span><br><span class="line"><span class="built_in">print</span>(<span class="string">f&#x27;Cells retained after scrublet: <span class="subst">&#123;n1&#125;</span>, <span class="subst">&#123;n0-n1&#125;</span> removed.&#x27;</span>)</span><br><span class="line"><span class="built_in">print</span>(<span class="string">f&#x27;End of post doublets removal and QC plots.&#x27;</span>)</span><br></pre></td></tr></table></figure>

<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">adata_manual=adata.copy()</span><br><span class="line">adata_auto=adata.copy()</span><br></pre></td></tr></table></figure>

<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line">tresh=&#123;<span class="string">&#x27;mito_perc&#x27;</span>: <span class="number">0.15</span>, <span class="string">&#x27;nUMIs&#x27;</span>: <span class="number">500</span>, <span class="string">&#x27;detected_genes&#x27;</span>: <span class="number">250</span>&#125;</span><br><span class="line"></span><br><span class="line">adata_manual.obs[<span class="string">&#x27;passing_mt&#x27;</span>] = adata_manual.obs[<span class="string">&#x27;mito_perc&#x27;</span>] &lt; tresh[<span class="string">&#x27;mito_perc&#x27;</span>]</span><br><span class="line">adata_manual.obs[<span class="string">&#x27;passing_nUMIs&#x27;</span>] = adata_manual.obs[<span class="string">&#x27;nUMIs&#x27;</span>] &gt; tresh[<span class="string">&#x27;nUMIs&#x27;</span>]</span><br><span class="line">adata_manual.obs[<span class="string">&#x27;passing_ngenes&#x27;</span>] = adata_manual.obs[<span class="string">&#x27;detected_genes&#x27;</span>] &gt; tresh[<span class="string">&#x27;detected_genes&#x27;</span>]</span><br><span class="line"></span><br><span class="line"><span class="built_in">print</span>(<span class="string">f&#x27;Lower treshold, nUMIs: <span class="subst">&#123;tresh[<span class="string">&quot;nUMIs&quot;</span>]&#125;</span>; filtered-out-cells: <span class="subst">&#123;n1-np.<span class="built_in">sum</span>(adata_manual.obs[<span class="string">&quot;passing_nUMIs&quot;</span>])&#125;</span>&#x27;</span>)</span><br><span class="line"><span class="built_in">print</span>(<span class="string">f&#x27;Lower treshold, n genes: <span class="subst">&#123;tresh[<span class="string">&quot;detected_genes&quot;</span>]&#125;</span>; filtered-out-cells: <span class="subst">&#123;n1-np.<span class="built_in">sum</span>(adata_manual.obs[<span class="string">&quot;passing_ngenes&quot;</span>])&#125;</span>&#x27;</span>)</span><br><span class="line"><span class="built_in">print</span>(<span class="string">f&#x27;Lower treshold, mito %: <span class="subst">&#123;tresh[<span class="string">&quot;mito_perc&quot;</span>]&#125;</span>; filtered-out-cells: <span class="subst">&#123;n1-np.<span class="built_in">sum</span>(adata_manual.obs[<span class="string">&quot;passing_mt&quot;</span>])&#125;</span>&#x27;</span>)</span><br></pre></td></tr></table></figure>

<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">QC_test = (adata_manual.obs[<span class="string">&#x27;passing_mt&#x27;</span>]) &amp; (adata_manual.obs[<span class="string">&#x27;passing_nUMIs&#x27;</span>]) &amp; (adata_manual.obs[<span class="string">&#x27;passing_ngenes&#x27;</span>])</span><br><span class="line">removed = QC_test.loc[<span class="keyword">lambda</span> x : x == <span class="literal">False</span>]</span><br><span class="line"><span class="built_in">print</span>(<span class="string">f&#x27;Total cell filtered out with this last  QC (and its chosen options): <span class="subst">&#123;n1-np.<span class="built_in">sum</span>(QC_test)&#125;</span>&#x27;</span>)</span><br><span class="line">adata_manual = adata_manual[QC_test, :].copy()</span><br><span class="line">n2 = adata_manual.shape[<span class="number">0</span>]</span><br><span class="line">    </span><br><span class="line"><span class="comment"># Store cleaned adata</span></span><br><span class="line"><span class="built_in">print</span>(<span class="string">f&#x27;Cells retained after scrublet and filtering: <span class="subst">&#123;n2&#125;</span>, <span class="subst">&#123;n0-n2&#125;</span> removed.&#x27;</span>)</span><br></pre></td></tr></table></figure>

<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">sc.pp.filter_cells(adata_manual, min_genes=<span class="number">200</span>)</span><br><span class="line">sc.pp.filter_genes(adata_manual, min_cells=<span class="number">3</span>)</span><br><span class="line"></span><br><span class="line">adata_manual</span><br></pre></td></tr></table></figure>

<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">adata=ov.pp.qc(adata,</span><br><span class="line">         tresh=&#123;<span class="string">&#x27;mito_perc&#x27;</span>: <span class="number">0.05</span>, <span class="string">&#x27;nUMIs&#x27;</span>: <span class="number">500</span>, <span class="string">&#x27;detected_genes&#x27;</span>: <span class="number">250</span>&#125;)</span><br><span class="line">adata</span><br></pre></td></tr></table></figure>

<h2 id="数据的标准化-归一化"><a href="#数据的标准化-归一化" class="headerlink" title="**数据的标准化&#x2F;**归一化"></a>**数据的标准化&#x2F;**<strong>归一化</strong></h2><p><img src="/post/single-cell/HVlqbME5woFXUuxEEDdcy168nEc.png"></p>
<p>“归一化”的预处理步骤旨在通过将“UMI 的方差”缩放到指定范围，来调整数据集中的原始 UMI 计数以实现模型建模。而在真实的单细胞分析中，有不同的归一化方法以解决不同的分析问题。</p>
<p>就是本章所要介绍的归一化内容，通过 benchmark 的测试，我们发现移位对数适用于大多数任务。但是如果我们的分析目标是寻找稀有细胞的时候，可以考虑采用皮尔森残差法来进行归一化。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">scales_counts = sc.pp.normalize_total(adata, target_sum=<span class="literal">None</span>, inplace=<span class="literal">False</span>)</span><br><span class="line"><span class="comment"># log1p transform</span></span><br><span class="line">adata.layers[<span class="string">&quot;log1p_norm&quot;</span>] = sc.pp.log1p(scales_counts[<span class="string">&quot;X&quot;</span>], copy=<span class="literal">True</span>)</span><br><span class="line"><span class="keyword">import</span> matplotlib.pyplot <span class="keyword">as</span> plt</span><br><span class="line">fig, axes = plt.subplots(<span class="number">1</span>, <span class="number">2</span>, figsize=(<span class="number">8</span>, <span class="number">4</span>))</span><br><span class="line">p1 = sns.histplot(adata.obs[<span class="string">&quot;total_counts&quot;</span>], bins=<span class="number">100</span>, kde=<span class="literal">False</span>, ax=axes[<span class="number">0</span>])</span><br><span class="line">axes[<span class="number">0</span>].set_title(<span class="string">&quot;Total counts&quot;</span>)</span><br><span class="line">p2 = sns.histplot(adata.layers[<span class="string">&quot;log1p_norm&quot;</span>].<span class="built_in">sum</span>(<span class="number">1</span>), bins=<span class="number">100</span>, kde=<span class="literal">False</span>, ax=axes[<span class="number">1</span>])</span><br><span class="line">axes[<span class="number">1</span>].set_title(<span class="string">&quot;Shifted logarithm&quot;</span>)</span><br><span class="line">plt.show()</span><br></pre></td></tr></table></figure>

<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> scipy.sparse <span class="keyword">import</span> csr_matrix</span><br><span class="line">analytic_pearson = sc.experimental.pp.normalize_pearson_residuals(adata, inplace=<span class="literal">False</span>)</span><br><span class="line">adata.layers[<span class="string">&quot;analytic_pearson_residuals&quot;</span>] = csr_matrix(analytic_pearson[<span class="string">&quot;X&quot;</span>])</span><br></pre></td></tr></table></figure>

<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">fig, axes = plt.subplots(<span class="number">1</span>, <span class="number">2</span>, figsize=(<span class="number">8</span>, <span class="number">4</span>))</span><br><span class="line">p1 = sns.histplot(adata.obs[<span class="string">&quot;total_counts&quot;</span>], bins=<span class="number">100</span>, kde=<span class="literal">False</span>, ax=axes[<span class="number">0</span>])</span><br><span class="line">axes[<span class="number">0</span>].set_title(<span class="string">&quot;Total counts&quot;</span>)</span><br><span class="line">p2 = sns.histplot(</span><br><span class="line">    adata.layers[<span class="string">&quot;analytic_pearson_residuals&quot;</span>].<span class="built_in">sum</span>(<span class="number">1</span>), bins=<span class="number">100</span>, kde=<span class="literal">False</span>, ax=axes[<span class="number">1</span>]</span><br><span class="line">)</span><br><span class="line">axes[<span class="number">1</span>].set_title(<span class="string">&quot;Analytic Pearson residuals&quot;</span>)</span><br><span class="line">plt.show()</span><br></pre></td></tr></table></figure>

<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">adata_pearson=ov.pp.preprocess(adata,mode=<span class="string">&#x27;pearson|pearson&#x27;</span>,n_HVGs=<span class="number">2000</span>,)</span><br><span class="line">adata_pearson</span><br></pre></td></tr></table></figure>

<h3 id="标准化"><a href="#标准化" class="headerlink" title="标准化"></a>标准化</h3><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 归一化，使得不同细胞样本间可比</span></span><br><span class="line"> sc.pp.normalize_total(adata, target_sum=<span class="number">1e4</span>)</span><br><span class="line"> sc.pp.log1p(adata)</span><br></pre></td></tr></table></figure>

<h3 id="将-AnnData-对象的-。raw-属性设置为归一化和对数化的原始基因表达（存储数据）"><a href="#将-AnnData-对象的-。raw-属性设置为归一化和对数化的原始基因表达（存储数据）" class="headerlink" title="将 AnnData 对象的 。raw 属性设置为归一化和对数化的原始基因表达（存储数据）"></a>将 AnnData 对象的 。raw 属性设置为归一化和对数化的原始基因表达（<strong>存储数据</strong>）</h3><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">adata.raw = adata</span><br></pre></td></tr></table></figure>

<p><strong>识别特异性基因（pl plotting（绘图））</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 计算</span></span><br><span class="line"> sc.pp.highly_variable_genes(adata, min_mean=<span class="number">0.0125</span>, max_mean=<span class="number">3</span>, min_disp=<span class="number">0.5</span>)</span><br><span class="line"><span class="comment">#min_mean: 最小均值阈值，用于筛选高度可变基因。默认值为0.0125。</span></span><br><span class="line"><span class="comment">#max_mean: 最大均值阈值，用于筛选高度可变基因。默认值为3。</span></span><br><span class="line"><span class="comment">#min_disp: 最小离散度阈值，用于筛选高度可变基因。默认值为0.5。</span></span><br><span class="line"><span class="comment">#n_top_genes：筛选出前n个高度可变基因。默认值为None，表示不进行筛选。</span></span><br><span class="line"><span class="comment">#n_bins：用于计算均值和离散度的箱数。默认值为20。</span></span><br><span class="line"><span class="comment">#subset: 子集索引，只在子集上计算高度可变基因。</span></span><br><span class="line"></span><br><span class="line"> <span class="comment"># 绘制特异性基因散点图</span></span><br><span class="line"> sc.pl.highly_variable_genes(adata)</span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/AKLfbLvxPoVbsoxLk71cGS6TnXb.png"></p>
<h3 id="获取只有特异性基因的数据集"><a href="#获取只有特异性基因的数据集" class="headerlink" title="获取只有特异性基因的数据集"></a><strong>获取只有特异性基因的数据集</strong></h3><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 获取只有特异性基因的数据集</span></span><br><span class="line"> adata = adata[:, adata.var.highly_variable]</span><br><span class="line"> <span class="comment"># 回归每个细胞的总计数和表达的线粒体基因的百分比的影响。</span></span><br><span class="line"> <span class="comment">#这一步的目的是消除总计数和线粒体基因表达对基因表达的影响，以减少这些因素在后续分析中的干扰。</span></span><br><span class="line"> sc.pp.regress_out(adata, [<span class="string">&#x27;total_counts&#x27;</span>, <span class="string">&#x27;pct_counts_mt&#x27;</span>])</span><br><span class="line"> <span class="comment"># 将每个基因缩放到单位方差。阈值超过标准偏差 10。</span></span><br><span class="line"> <span class="comment">#这一步的目的是对基因表达进行归一化，使得不同基因具有相同的方差，</span></span><br><span class="line"> <span class="comment">#并将表达值限制在一定范围内，以便后续分析中的比较和可视化更具可解释性。</span></span><br><span class="line"> sc.pp.scale(adata, max_value=<span class="number">10</span>)</span><br><span class="line"> <span class="comment">#layer：要使用的数据层。默认为 None，表示使用原始的基因表达数据。你可以指定其他层的名称，例如批次校正后的数据层。</span></span><br><span class="line"><span class="comment">#obsm：要使用的观测值特征矩阵。默认为 None。可以指定一个观测值特征矩阵来进行缩放，例如使用 UMAP 嵌入作为观测值特征矩阵来进行缩放。</span></span><br><span class="line"><span class="comment">#varm：要使用的变量特征矩阵。默认为 None。可以指定一个变量特征矩阵来进行缩放，例如使用基因的重要性分数作为变量特征矩阵来进行缩放。</span></span><br></pre></td></tr></table></figure>

<h3 id="如何评估细胞周期的影响？"><a href="#如何评估细胞周期的影响？" class="headerlink" title="如何评估细胞周期的影响？"></a>如何评估细胞周期的影响？</h3><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">sc.tl.scores_genes_cell_cycle(adata,s_genes=s_genes,g2m_genes=g2m_genes)</span><br><span class="line"><span class="comment">#s_genes：包含 G1/S 相转录本的基因列表，可以是基因名称的字符串列表或布尔索引。默认为 None。</span></span><br><span class="line"><span class="comment">#g2m_genes：包含 G2/M 相转录本的基因列表，可以是基因名称的字符串列表或布尔索引。默认为 None</span></span><br></pre></td></tr></table></figure>

<p>函数用于计算每个细胞的细胞周期得分，根据指定的 G1&#x2F;S 和 G2&#x2F;M 基因集。</p>
<p>函数会计算并添加以下列到 <code>adata.obs</code> 中：</p>
<ul>
<li><code>&#39;phase&#39;</code>：细胞周期阶段的注释，根据得分进行划分为 ‘G1’， ‘S’， ‘G2M’ 和 ‘unknown’。</li>
<li><code>&#39;G1_score&#39;</code>：G1&#x2F;S 得分。</li>
<li><code>&#39;S_score&#39;</code>：S 得分。</li>
<li><code>&#39;G2M_score&#39;</code>：G2&#x2F;M 得分。</li>
</ul>
<p>使用时，需要提供适当的 G1&#x2F;S 和 G2&#x2F;M 基因列表，或者可以使用预定义的基因集。函数会根据基因表达数据计算细胞周期得分，并为每个细胞分配细胞周期阶段的注释。</p>
<h2 id="主成分分析-（数据降维）"><a href="#主成分分析-（数据降维）" class="headerlink" title="主成分分析****（数据降维）"></a><strong>主成分分析****（数据降维）</strong></h2><h3 id="通过运行主成分分析-（PCA）-来降低数据的维数，可以对数据进行去噪并揭示不同分群的主因素。"><a href="#通过运行主成分分析-（PCA）-来降低数据的维数，可以对数据进行去噪并揭示不同分群的主因素。" class="headerlink" title="通过运行主成分分析 （PCA） 来降低数据的维数，可以对数据进行去噪并揭示不同分群的主因素。"></a>通过运行主成分分析 （PCA） 来降低数据的维数，可以对数据进行去噪并揭示不同分群的主因素。</h3><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 绘制 PCA 图</span></span><br><span class="line"> sc.pl.pca(adata, color=<span class="string">&#x27;CST3&#x27;</span>)</span><br><span class="line"> <span class="comment">#color：用于着色的变量名称或基因名称。可以是存储在 adata.obs 或 adata.var 中的变量名，</span></span><br><span class="line"> <span class="comment">#或是直接传入基因名称的字符串。默认为 None</span></span><br><span class="line"> <span class="comment">#projection：指定要使用的降维方法。可以是字符串（如 &#x27;pca&#x27;、&#x27;tsne&#x27;、&#x27;umap&#x27;）或自定义降维函数。</span></span><br><span class="line"> <span class="comment">#默认为 &#x27;pca&#x27;</span></span><br><span class="line"> <span class="comment">#svd_solver 参数用于指定奇异值分解（SVD）的求解器。SVD 是 PCA 的一种实现方法。该参数的可选值包括：</span></span><br><span class="line"><span class="comment">#&#x27;auto&#x27;：根据数据的大小和特征数量自动选择求解器。对于大型数据集，通常选择 &#x27;randomized&#x27; 求解器。</span></span><br><span class="line"><span class="comment">#&#x27;arpack&#x27;：使用 ARPACK 求解器进行 SVD。适用于中小型数据集。</span></span><br><span class="line"><span class="comment">#&#x27;randomized&#x27;：使用随机 SVD 求解器进行 SVD。适用于大型数据集。</span></span><br><span class="line"><span class="comment">#通过设置 svd_solver 参数，可以根据数据集的大小和性质选择适当的求解器.</span></span><br><span class="line"><span class="comment">#默认值为 &#x27;auto&#x27;，会根据数据自动选择最合适的求解器。</span></span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/VNWPbvl2LoRuREx6PaLcA1vznGh.png"></p>
<h3 id="检查单个-PC-对数据总方差的贡献，这可以提供给我们应该考虑多少个-PC-以计算细胞的邻域关系的信息，"><a href="#检查单个-PC-对数据总方差的贡献，这可以提供给我们应该考虑多少个-PC-以计算细胞的邻域关系的信息，" class="headerlink" title="检查单个 PC 对数据总方差的贡献，这可以提供给我们应该考虑多少个 PC 以计算细胞的邻域关系的信息，"></a>检查单个 PC 对数据总方差的贡献，这可以提供给我们应该考虑多少个 PC 以计算细胞的邻域关系的信息，</h3><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">sc.pl.pca_variance_ratio(adata, log=<span class="literal">True</span>)</span><br><span class="line">** <span class="comment">#大概哪里下降的缓慢一些，越平计算量越大**</span></span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/BKgSbhuZ3o7JN6xuSyxcGciyn6f.png"></p>
<h2 id="细胞分群"><a href="#细胞分群" class="headerlink" title="细胞分群"></a><strong>细胞分群</strong></h2><p><img src="/post/single-cell/XWdUbqx1IovarAxZGOac982rn9e.png"></p>
<p>使用数据矩阵的 PCA 表示来计算细胞的邻域图。</p>
<p><strong>建议使用 UMAP ，它可能比 tSNE 更忠实于<strong><strong>流形</strong></strong>的全局连通性，因此能更好地保留轨迹。</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">sc.pp.neighbors(adata, n_neighbors=<span class="number">10</span>, n_pcs=<span class="number">40</span>)</span><br><span class="line"><span class="comment">#n_neighbors：指定每个细胞的近邻数量，默认为 15。较大的值可以提高邻居关系的准确性，但也增加了计算时间和内存消耗。</span></span><br><span class="line"><span class="comment">#n_pcs：指定要保留的主成分数量，默认为 50。主成分用于计算细胞之间的相似性，并用于构建邻居关系图。根据自己选择的pc</span></span><br><span class="line"><span class="comment">#random_state：用于控制随机数生成器的种子，以确保可复现性。</span></span><br><span class="line"> sc.tl.umap(adata)</span><br><span class="line"> <span class="comment"># 如果设置了 adata 的 .raw 属性时，下图显示了“raw”（标准化、对数化但未校正）基因表达矩阵。</span></span><br><span class="line"> sc.pl.umap(adata, color=[<span class="string">&#x27;CST3&#x27;</span>, <span class="string">&#x27;NKG7&#x27;</span>, <span class="string">&#x27;PPBP&#x27;</span>])</span><br><span class="line"> <span class="comment">#您可以通过设置 color 参数来指定在图中显示的基因。在示例中，基因 &#x27;CST3&#x27;、&#x27;NKG7&#x27; 和 &#x27;PPBP&#x27; 被选择为要在 UMAP 图中显示的颜色标记。</span></span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/SZNvb6PvsoVbsYxJxuHcttlXnCe.png"></p>
<p>为了绘制缩放矫正的基因表达聚类图，需要使用 use_raw&#x3D;False 参数。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">sc.pl.umap(adata, color=[<span class="string">&#x27;CST3&#x27;</span>, <span class="string">&#x27;NKG7&#x27;</span>, <span class="string">&#x27;PPBP&#x27;</span>], use_raw=<span class="literal">False</span>)</span><br><span class="line"> <span class="comment">#use_raw 参数设置为 False。这将确保在可视化过程中使用缩放校正后的数据。</span></span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/KbQmbVDhyojvQMx2uufcTlMLnYc.png"></p>
<p>目前还没有计算出各个细胞类群，下面进行聚类：</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 计算</span></span><br><span class="line"> sc.tl.leiden(adata)</span><br><span class="line"> <span class="comment"># 绘制</span></span><br><span class="line"> sc.pl.umap(adata, color=[<span class="string">&#x27;leiden&#x27;</span>])</span><br><span class="line"> </span><br><span class="line"> sc.tl.leiden(adata, resolution=<span class="number">0.5</span>, n_neighbors=<span class="number">20</span>, key_added=<span class="string">&#x27;cluster&#x27;</span>)</span><br><span class="line"><span class="comment">#resolution：用于调节聚类的分辨率参数，值越大会产生更多的聚类簇，默认为1.0。</span></span><br><span class="line"><span class="comment">#n_neighbors：用于计算最近邻的邻居数量，默认为30。</span></span><br><span class="line"><span class="comment">#random_state：用于控制随机数生成器的种子，以确保可复现性。</span></span><br><span class="line"><span class="comment">#key_added：用于指定存储聚类结果的关键字，将聚类结果存储在adata.obs中，默认为leiden。</span></span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/TdYTbYEwMoIAHLx8iOucWuxInLt.png"></p>
<h2 id="检索生物学标记基因"><a href="#检索生物学标记基因" class="headerlink" title="检索生物学标记基因"></a><strong>检索生物学标记基因</strong></h2><p>先计算每个 leiden 分群中高度差异基因的排名，取排名前 25 的基因。</p>
<p>默认情况下，使用 AnnData 的 。raw 属性。</p>
<p>在基因组学分析中，常用的方法之一是寻找不同条件下的差异表达基因。不同的差异表达分析方法适用于不同的数据类型和假设条件，具有不同的统计原理和计算方法。</p>
<ul>
<li>T-检验（t-test）：适用于两组样本（例如，疾病组和对照组）之间的差异分析。它基于样本均值之间的差异和方差估计，检验是否存在显著差异。T-检验假设数据符合正态分布。</li>
<li>Wilcoxon 秩和检验（Wilcoxon）：也称为 Mann-Whitney U 检验，适用于两组样本之间的差异分析。它基于将两组样本的值合并，并对合并的值进行排序，然后计算每个样本的秩和。检验是否存在显著差异。Wilcoxon 检验不依赖于数据的分布，更适合非正态分布的数据。</li>
<li>逻辑回归（logreg）：适用于多组样本（例如，多个疾病亚型之间的差异分析）。逻辑回归使用分类模型，将样本分为多个类别，并计算每个基因在不同类别之间的差异。逻辑回归可以考虑多个因素对差异的影响，并给出基因在各个类别中的表达模式。</li>
</ul>
<p>选择使用哪种方法取决于数据类型、实验设计和研究问题的特点。在实际应用中，可以尝试多种方法，并根据结果进行综合分析和解释。</p>
<h3 id="T-test"><a href="#T-test" class="headerlink" title="T-test"></a><strong>T-test</strong></h3><p>最简单和最快的方法是 t 检验。tl tools</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">sc.tl.rank_genes_groups(adata, <span class="string">&#x27;leiden&#x27;</span>, method=<span class="string">&#x27;t-test&#x27;</span>)</span><br><span class="line"> <span class="comment">#基于leiden的分群</span></span><br><span class="line"> <span class="comment">#rankby_abs=True,按差异的绝对值大小进行排序</span></span><br><span class="line"><span class="comment">#reference=&#x27;1&#x27;: 可选参数，用于指定参考组。默认为 None，表示使用所有组进行比较。</span></span><br><span class="line"><span class="comment">#groups=&#x27;0&#x27;: 可选参数，用于指定要比较的组。默认为 None，表示使用所有组进行比较。</span></span><br><span class="line"> </span><br><span class="line"> sc.pl.rank_genes_groups(adata, n_genes=<span class="number">25</span>, sharey=<span class="literal">False</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/Oi2obWcbNoxGBTxX6P5cXYjLnkf.png"></p>
<p><img src="/post/single-cell/CC9ubhzlco0u7TxnKzPcz2GPnee.png"></p>
<h3 id="Wilcoxon-rank-sum"><a href="#Wilcoxon-rank-sum" class="headerlink" title="Wilcoxon rank-sum"></a><strong>Wilcoxon rank-sum</strong></h3><p>Wilcoxon rank-sum （Mann-Whitney-U） 检验的结果非常相似，还可以使用其他的差异分析包，如 MAST、limma、DESeq2 和 diffxpy。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">sc.tl.rank_genes_groups(adata, <span class="string">&#x27;leiden&#x27;</span>, method=<span class="string">&#x27;wilcoxon&#x27;</span>)</span><br><span class="line"> sc.pl.rank_genes_groups(adata, n_genes=<span class="number">25</span>, sharey=<span class="literal">False</span>)</span><br><span class="line"> <span class="comment"># 保存这次的数据结果</span></span><br><span class="line"> adata.write(results_file)</span><br><span class="line"> <span class="comment">#请注意，上述代码片段中的 results_file 是指定的文件路径，您需要根据需要自行指定保存文件的位置和文件名。</span></span><br></pre></td></tr></table></figure>

<h3 id="逻辑回归"><a href="#逻辑回归" class="headerlink" title="逻辑回归"></a><strong>逻辑回归</strong></h3><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">sc.tl.rank_genes_groups(adata, <span class="string">&#x27;leiden&#x27;</span>, method=<span class="string">&#x27;logreg&#x27;</span>)</span><br><span class="line"> sc.pl.rank_genes_groups(adata, n_genes=<span class="number">25</span>, sharey=<span class="literal">False</span>)</span><br><span class="line"> <span class="comment">#显示每个聚类簇中具有最显著差异的前25个基因</span></span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/SsL6boVvOoqtL5xxccmcJ2Y9nKg.png"></p>
<h3 id="载入数据"><a href="#载入数据" class="headerlink" title="载入数据"></a><strong>载入数据</strong></h3><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 使用 Wilcoxon Rank-Sum 测试结果重新加载已保存的对象</span></span><br><span class="line"> adata = sc.read(results_file)</span><br></pre></td></tr></table></figure>

<h3 id="获取聚类分组和分数"><a href="#获取聚类分组和分数" class="headerlink" title="获取聚类分组和分数"></a><strong>获取聚类分组和分数</strong></h3><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">result = adata.uns[<span class="string">&#x27;rank_genes_groups&#x27;</span>]</span><br><span class="line"> groups = result[<span class="string">&#x27;names&#x27;</span>].dtype.names</span><br><span class="line"> pd.DataFrame(</span><br><span class="line">     &#123;group + <span class="string">&#x27;_&#x27;</span> + key[:<span class="number">1</span>]: result[key][group]</span><br><span class="line">     <span class="keyword">for</span> group <span class="keyword">in</span> groups <span class="keyword">for</span> key <span class="keyword">in</span> [<span class="string">&#x27;names&#x27;</span>, <span class="string">&#x27;pvals_adj&#x27;</span>]&#125;).head(<span class="number">5</span>)</span><br><span class="line">     </span><br><span class="line">这段代码主要用于提取差异表达分析结果中的关键信息并创建一个包含前<span class="number">5</span>个差异基因的DataFrame。</span><br><span class="line">首先，从adata.uns[<span class="string">&#x27;rank_genes_groups&#x27;</span>]中获取差异表达分析的结果，存储在变量result中。该结果通常包含不同组或簇之间的差异基因信息。</span><br><span class="line">接下来，通过result[<span class="string">&#x27;names&#x27;</span>].dtype.names获取所有组或簇的名称，存储在变量groups中。这些组或簇的名称在差异表达分析中用于对基因进行分类或分组。</span><br><span class="line">然后，使用字典推导式创建一个字典，其中键是组或簇的名称加上<span class="string">&#x27;name&#x27;</span>或<span class="string">&#x27;pvals&#x27;</span>的首字母，值是对应的结果数据。这个字典的目的是构建DataFrame的列。</span><br><span class="line">最后，使用pd.DataFrame()函数将字典转换为DataFrame，并只选择前<span class="number">5</span>个差异基因进行展示。生成的DataFrame包含了每个组或簇的差异基因名称和对应的p-value。</span><br><span class="line">通过这段代码，可以方便地查看差异表达分析结果中的关键信息，包括差异基因名称和p-value，并以DataFrame的形式展示出来，便于后续分析和可视化。</span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/GGasbq05PoaPZOxTYJvcrmfUnQg.png"></p>
<h2 id="重要方法"><a href="#重要方法" class="headerlink" title="重要方法"></a>重要方法</h2><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">result = adata.uns[<span class="string">&#x27;rank_genes_groups&#x27;</span>]</span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/KrNYbErPDooPdmxiCKCcK1W7nhf.png"></p>
<p><img src="/post/single-cell/JDgNblbuRoth4wx124Yc4CL8nWd.png"></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">adata.uns[<span class="string">&#x27;rank_genes_groups&#x27;</span>][<span class="string">&#x27;names&#x27;</span>][<span class="string">&#x27;0&#x27;</span>]</span><br><span class="line"><span class="comment">#第一列</span></span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/VUYKbUvWZoohAzxPhnFc6qSAnch.png"></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"></span><br><span class="line"><span class="comment"># 创建初始的空数组，形状为(1, 5)</span></span><br><span class="line">out = np.array([<span class="number">0</span>, <span class="number">0</span>, <span class="number">0</span>, <span class="number">0</span>, <span class="number">0</span>])</span><br><span class="line"></span><br><span class="line"><span class="comment"># 遍历每个群组</span></span><br><span class="line"><span class="keyword">for</span> group <span class="keyword">in</span> results[<span class="string">&#x27;names&#x27;</span>].dtype.names:</span><br><span class="line">    <span class="comment"># 获取当前群组的基因名称、得分、P值、对数折变值</span></span><br><span class="line">    gene_names = results[<span class="string">&#x27;names&#x27;</span>][group]</span><br><span class="line">    scores = results[<span class="string">&#x27;scores&#x27;</span>][group]</span><br><span class="line">    pvals = results[<span class="string">&#x27;pvals_adf&#x27;</span>][group]</span><br><span class="line">    logfoldchanges = results[<span class="string">&#x27;logfoldchanges&#x27;</span>][group]</span><br><span class="line">    </span><br><span class="line">    <span class="comment"># 创建与基因名称相同长度的数组，每个元素为当前群组的名称</span></span><br><span class="line">    group_names = np.array([group] * <span class="built_in">len</span>(gene_names)).astype(<span class="string">&#x27;object&#x27;</span>)</span><br><span class="line">    </span><br><span class="line">    <span class="comment"># 将基因名称、得分、P值、对数折变值和群组名称的数组进行转置</span></span><br><span class="line">    group_data = np.vstack((gene_names, scores, pvals, logfoldchanges, group_names)).T</span><br><span class="line">    </span><br><span class="line">    <span class="comment"># 将当前群组的结果垂直堆叠到初始的空数组中</span></span><br><span class="line">    out = np.vstack((out, group_data))</span><br><span class="line"></span><br><span class="line"><span class="comment"># 去除初始的空数组，并得到最终的结果数组</span></span><br><span class="line">markes=pd.DataFrame(out[<span class="number">1</span>:],columns=[<span class="string">&#x27;Gene&#x27;</span>,<span class="string">&#x27;scores&#x27;</span>,<span class="string">&#x27;pvals_adj&#x27;</span>,<span class="string">&#x27;lfc&#x27;</span>,<span class="string">&#x27;cluster&#x27;</span>])</span><br><span class="line"></span><br><span class="line"><span class="comment"># 注释：在这里，`out`将包含所有群组的基因名称、得分、P值、对数折变值和群组名称的信息</span></span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/R7d8bVISkoNZKyxEwvzcmpEEnxb.png"></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">markes=markers[(markes.pval_adj&lt;<span class="number">0.05</span>) &amp;(<span class="built_in">abs</span>(markers.lfc)&gt;<span class="number">1</span>)]</span><br><span class="line"><span class="comment">#在差异表达分析中，&quot;lfc&quot;代表&quot;log-fold change&quot;，是一种衡量基因表达差异大小的指标。它表示在两个条件或组之间的基因表达水平的倍数差异的对数值。正的lfc表示基因在条件A中的表达水平高于条件B，负的lfc表示基因在条件B中的表达水平高于条件A。lfc的绝对值越大，表示差异越显著。</span></span><br><span class="line"><span class="comment">#&quot;pval_adj&quot;代表校正后的p-value（经过多重假设检验校正后的p-value），也称为调整后的显著性水平。在差异表达分析中，p-value用于评估差异是否统计上显著。一般情况下，p-value小于设定的显著性水平（通常为0.05）被认为是显著差异的阈值。然而，由于进行多重假设检验时可能出现假阳性的问题，需要对p-value进行校正，以控制整体错误率。校正后的p-value（pval_adj）考虑了多重比较的问题，提供了更可靠的显著性评估。较小的pval_adj值表示差异更显著。</span></span><br><span class="line"><span class="comment">#在给定的上下文中，lfc和pval_adj是差异表达分析中常用的指标，用于评估基因在不同条件或组之间的差异程度和显著性水平。</span></span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/QHVjbcNolodhKbxUrHJccDTlnsd.png"></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">markers=markers[markers.cluser==<span class="string">&#x27;4&#x27;</span>]</span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/UilzbcvrJopipkxl8ijcCgoKn3b.png"></p>
<h2 id="找到特定的基因-zika"><a href="#找到特定的基因-zika" class="headerlink" title="找到特定的基因-zika"></a>找到特定的基因-zika</h2><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 获取基因名为&#x27;zika&#x27;的索引位置</span></span><br><span class="line">zika_i = np.where(adata.raw.var_names == <span class="string">&#x27;zika&#x27;</span>)[<span class="number">0</span>][<span class="number">0</span>]</span><br><span class="line"></span><br><span class="line"><span class="comment"># 提取基因&#x27;zika&#x27;的表达数据</span></span><br><span class="line">zika = adata.raw.X.toarray()[:, zika_i]</span><br><span class="line"></span><br><span class="line"><span class="comment"># 将基因&#x27;zika&#x27;的表达数据添加为新的观测列&#x27;Zika&#x27;</span></span><br><span class="line">adata.obs[<span class="string">&#x27;Zika&#x27;</span>] = zika</span><br><span class="line"></span><br><span class="line"><span class="comment"># 根据&#x27;Zika&#x27;列的值大于0进行过滤，仅保留满足条件的观测</span></span><br><span class="line">filtered_obs = adata.obs[adata.obs[<span class="string">&#x27;Zika&#x27;</span>] &gt; <span class="number">0</span>]</span><br><span class="line"></span><br><span class="line"><span class="comment"># 使用sc.pl.violin函数绘制基因&#x27;zika&#x27;在不同群集(leiden)中的小提琴图</span></span><br><span class="line">sc.pl.violin(adata, [<span class="string">&#x27;zika&#x27;</span>], groupby=<span class="string">&#x27;leiden&#x27;</span>, use_raw=<span class="literal">True</span>)</span><br><span class="line"><span class="comment">#其中use_raw=True表示使用原始数据进行绘图。</span></span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/R08abvpB5oj2M3xIS4BcjKjBnR4.png"></p>
<h3 id="Group-1-与-Group-2-比较，进行差异分析"><a href="#Group-1-与-Group-2-比较，进行差异分析" class="headerlink" title="Group 1 与 Group 2 比较，进行差异分析"></a><strong>Group 1 与 Group 2 比较，进行差异分析</strong></h3><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">sc.tl.rank_genes_groups(adata, <span class="string">&#x27;leiden&#x27;</span>, groups=[<span class="string">&#x27;0&#x27;</span>], reference=<span class="string">&#x27;1&#x27;</span>, method=<span class="string">&#x27;wilcoxon&#x27;</span>)</span><br><span class="line"> sc.pl.rank_genes_groups(adata, groups=[<span class="string">&#x27;0&#x27;</span>], n_genes=<span class="number">20</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/MMXrbSQwVocWTxxsyU6c9CCdnSe.png"></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sc.pl.rank_genes_groups_violin(adata, groups=<span class="string">&#x27;0&#x27;</span>, n_genes=<span class="number">8</span>)</span><br></pre></td></tr></table></figure>

<p>一分为二</p>
<p><img src="/post/single-cell/HPJCbZws9oKvJJxQfU8cu0pKnoe.png"></p>
<h3 id="Group-0-与其余组的比较进行差异分析"><a href="#Group-0-与其余组的比较进行差异分析" class="headerlink" title="Group 0 与其余组的比较进行差异分析"></a><strong>Group 0 与其余组的比较进行差异分析</strong></h3><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">adata = sc.read(results_file)</span><br><span class="line"> sc.pl.rank_genes_groups_violin(adata, groups=<span class="string">&#x27;0&#x27;</span>, n_genes=<span class="number">8</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/RSpNbdZpHoEe4fxcIivcueCyn7b.png"></p>
<h3 id="跨类群比较基因"><a href="#跨类群比较基因" class="headerlink" title="跨类群比较基因"></a><strong>跨类群比较基因</strong></h3><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sc.pl.violin(adata, [<span class="string">&#x27;CST3&#x27;</span>, <span class="string">&#x27;NKG7&#x27;</span>, <span class="string">&#x27;PPBP&#x27;</span>], groupby=<span class="string">&#x27;leiden&#x27;</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/LA4VbrHVxol5JZxGzdhcf1nVnfg.png"></p>
<h2 id="根据已知的细胞标记，注释细胞类型"><a href="#根据已知的细胞标记，注释细胞类型" class="headerlink" title="根据已知的细胞标记，注释细胞类型"></a><strong>根据已知的细胞标记，注释细胞类型</strong></h2><h4 id="法一："><a href="#法一：" class="headerlink" title="法一："></a>法一：</h4><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">new_cluster_names = [</span><br><span class="line">     <span class="string">&#x27;CD4 T&#x27;</span>, <span class="string">&#x27;CD14 Monocytes&#x27;</span>,</span><br><span class="line">     <span class="string">&#x27;B&#x27;</span>, <span class="string">&#x27;CD8 T&#x27;</span>,</span><br><span class="line">     <span class="string">&#x27;NK&#x27;</span>, <span class="string">&#x27;FCGR3A Monocytes&#x27;</span>,</span><br><span class="line">     <span class="string">&#x27;Dendritic&#x27;</span>, <span class="string">&#x27;Megakaryocytes&#x27;</span>]</span><br><span class="line"> adata.rename_categories(<span class="string">&#x27;leiden&#x27;</span>, new_cluster_names)</span><br><span class="line"> sc.pl.umap(adata, color=<span class="string">&#x27;leiden&#x27;</span>, legend_loc=<span class="string">&#x27;on data&#x27;</span>, title=<span class="string">&#x27;&#x27;</span>, frameon=<span class="literal">False</span>, save=<span class="string">&#x27;.pdf&#x27;</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/ShA8bZdyuok0sOxCqaecBPWBnCb.png"></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 根据已知的标记基因，定义一个标记基因列表供以后参考：</span></span><br><span class="line"> marker_genes = [<span class="string">&#x27;IL7R&#x27;</span>, <span class="string">&#x27;CD79A&#x27;</span>, <span class="string">&#x27;MS4A1&#x27;</span>, <span class="string">&#x27;CD8A&#x27;</span>, <span class="string">&#x27;CD8B&#x27;</span>, <span class="string">&#x27;LYZ&#x27;</span>, <span class="string">&#x27;CD14&#x27;</span>,</span><br><span class="line">                 <span class="string">&#x27;LGALS3&#x27;</span>, <span class="string">&#x27;S100A8&#x27;</span>, <span class="string">&#x27;GNLY&#x27;</span>, <span class="string">&#x27;NKG7&#x27;</span>, <span class="string">&#x27;KLRB1&#x27;</span>,</span><br><span class="line">                 <span class="string">&#x27;FCGR3A&#x27;</span>, <span class="string">&#x27;MS4A7&#x27;</span>, <span class="string">&#x27;FCER1A&#x27;</span>, <span class="string">&#x27;CST3&#x27;</span>, <span class="string">&#x27;PPBP&#x27;</span>]</span><br><span class="line">                 </span><br><span class="line"> sc.pl.dotplot(adata, marker_genes, groupby=<span class="string">&#x27;leiden&#x27;</span>)</span><br><span class="line"> <span class="comment">#sc.pl.dotplot() 函数用于绘制基因的点图，展示它们在不同细胞群组中的表达情况。</span></span><br><span class="line"> <span class="comment">#use_raw=True,       # 是否使用原始（未处理）的基因表达数据，默认为 True</span></span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/U8zcbHFvvoWNAcx9xiicghl9nqd.png"></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sc.pl.stacked_violin(adata, marker_genes, groupby=<span class="string">&#x27;leiden&#x27;</span>, rotation=<span class="number">90</span>);</span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/YSinbm2ZNoaUzXxGtkGc6TDFnYe.png"></p>
<h4 id="法二："><a href="#法二：" class="headerlink" title="法二："></a>法二：</h4><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 根据已知的标记基因，定义一个标记基因列表供以后参考：</span></span><br><span class="line"> marker_genes = [<span class="string">&#x27;IL7R&#x27;</span>, <span class="string">&#x27;CD79A&#x27;</span>, <span class="string">&#x27;MS4A1&#x27;</span>, <span class="string">&#x27;CD8A&#x27;</span>, <span class="string">&#x27;CD8B&#x27;</span>, <span class="string">&#x27;LYZ&#x27;</span>, <span class="string">&#x27;CD14&#x27;</span>,</span><br><span class="line">                 <span class="string">&#x27;LGALS3&#x27;</span>, <span class="string">&#x27;S100A8&#x27;</span>, <span class="string">&#x27;GNLY&#x27;</span>, <span class="string">&#x27;NKG7&#x27;</span>, <span class="string">&#x27;KLRB1&#x27;</span>,</span><br><span class="line">                 <span class="string">&#x27;FCGR3A&#x27;</span>, <span class="string">&#x27;MS4A7&#x27;</span>, <span class="string">&#x27;FCER1A&#x27;</span>, <span class="string">&#x27;CST3&#x27;</span>, <span class="string">&#x27;PPBP&#x27;</span>]</span><br><span class="line">                 </span><br><span class="line"> sc.pl.dotplot(adata, marker_genes, groupby=<span class="string">&#x27;leiden&#x27;</span>)</span><br><span class="line"> <span class="comment">#sc.pl.dotplot() 函数用于绘制基因的点图，展示它们在不同细胞群组中的表达情况。</span></span><br><span class="line"> <span class="comment">#use_raw=True,       # 是否使用原始（未处理）的基因表达数据，默认为 True</span></span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/TENVbeNlPo3mm8xrbgXc18vQnbw.png"></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> matplotlib.pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"></span><br><span class="line">new_cluster_names = [<span class="string">&#x27;CD4 T&#x27;</span>, <span class="string">&#x27;CD14 Monocytes&#x27;</span>, <span class="string">&#x27;B&#x27;</span>, <span class="string">&#x27;CD8 T&#x27;</span>, <span class="string">&#x27;FCGR3A Monocytes&#x27;</span>, <span class="string">&#x27;NK&#x27;</span>, <span class="string">&#x27;Dendritic&#x27;</span>, <span class="string">&#x27;Megakaryocytes&#x27;</span>]</span><br><span class="line">new_cluster_names_unique = <span class="built_in">list</span>(<span class="built_in">set</span>(new_cluster_names))</span><br><span class="line">new_cluster_names_unique.sort(key=new_cluster_names.index)</span><br><span class="line"></span><br><span class="line">celltype = [new_cluster_names_unique[new_cluster_names.index(i)] <span class="keyword">for</span> i <span class="keyword">in</span> adata.obs.leiden.values.tolist()]</span><br><span class="line">adata.obs[<span class="string">&#x27;leiden&#x27;</span>] = celltype</span><br><span class="line">adata.obs[<span class="string">&#x27;leiden&#x27;</span>] = adata.obs[<span class="string">&#x27;leiden&#x27;</span>].astype(<span class="string">&#x27;category&#x27;</span>)</span><br><span class="line">adata.obs[<span class="string">&#x27;leiden&#x27;</span>].cat.set_categories(new_cluster_names_unique, inplace=<span class="literal">True</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">with</span> plt.rc_context(&#123;<span class="string">&#x27;figure.figsize&#x27;</span>: (<span class="number">10</span>, <span class="number">10</span>)&#125;):</span><br><span class="line">    sc.pl.umap(adata, color=<span class="string">&#x27;leiden&#x27;</span>, legend_loc=<span class="string">&#x27;on data&#x27;</span>, title=<span class="string">&#x27;&#x27;</span>, frameon=<span class="literal">True</span>, legend_fontsize=<span class="string">&#x27;x-small&#x27;</span>)</span><br><span class="line">    plt.savefig(<span class="string">&#x27;plot/celltype_umap.pdf&#x27;</span>)</span><br><span class="line"></span><br><span class="line">sc.pl.dotplot(adata, marker_genes, groupby=<span class="string">&#x27;leiden&#x27;</span>)</span><br><span class="line">plt.savefig(<span class="string">&#x27;plot/celltype_marker_dotplot.pdf&#x27;</span>)</span><br><span class="line"></span><br><span class="line">sc.pl.stacked_violin(adata, marker_genes, groupby=<span class="string">&#x27;leiden&#x27;</span>, rotation=<span class="number">90</span>)</span><br><span class="line">plt.savefig(<span class="string">&#x27;plot/celltype_marker_violin.pdf&#x27;</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/LlFUbIbuXoiS0AxrSXSc4wO3ndb.png"></p>
<h1 id="数据保存"><a href="#数据保存" class="headerlink" title="数据保存"></a>数据保存</h1><p>如果只想将其用于可视化的人共享此文件，减少文件大小的一种简单方法是删除缩放和校正的数据矩阵。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">adata.write(results_file, compression=<span class="string">&#x27;gzip&#x27;</span>)</span><br></pre></td></tr></table></figure>

<p><strong>保存为 h5ad 数据</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">adata.raw.to_adata().write(<span class="string">&#x27;./write/pbmc3k_withoutX.h5ad&#x27;</span>)</span><br></pre></td></tr></table></figure>

<p><strong>读取 h5ad 数据</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">adata = sc.read_h5ad(<span class="string">&#x27;./write/pbmc3k_withoutX.h5ad&#x27;</span>)</span><br></pre></td></tr></table></figure>

<p><strong>导出数据子集</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 导出聚类数据</span></span><br><span class="line"> adata.obs[[<span class="string">&#x27;n_counts&#x27;</span>, <span class="string">&#x27;louvain_groups&#x27;</span>]].to_csv(<span class="string">&#x27;./write/pbmc3k_corrected_louvain_groups.csv&#x27;</span>)</span><br><span class="line"> <span class="comment"># 导出PCA数据</span></span><br><span class="line"> adata.obsm.to_df()[[<span class="string">&#x27;X_pca1&#x27;</span>, <span class="string">&#x27;X_pca2&#x27;</span>]].to_csv(<span class="string">&#x27;./write/pbmc3k_corrected_X_pca.csv&#x27;</span>)</span><br></pre></td></tr></table></figure>

<h1 id="可视化"><a href="#可视化" class="headerlink" title="可视化"></a>可视化</h1><h2 id="基因表达"><a href="#基因表达" class="headerlink" title="基因表达"></a>基因表达</h2><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> re</span><br><span class="line"><span class="keyword">from</span> matplotlib <span class="keyword">import</span> pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">from</span> matplotlib.pyplot <span class="keyword">import</span> rc_context</span><br><span class="line"></span><br><span class="line"><span class="comment"># 设置Scanpy绘图参数</span></span><br><span class="line">sc.set_figure_params(dpi=<span class="number">100</span>, color_map=<span class="string">&#x27;viridis_r&#x27;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 设置Scanpy的日志显示级别</span></span><br><span class="line">sc.settings.verbosity = <span class="number">1</span></span><br><span class="line">sc.logging.print_header()</span><br><span class="line"></span><br><span class="line"><span class="comment"># 手动下载数据集</span></span><br><span class="line"><span class="comment">#绝对路径</span></span><br><span class="line">data_path = <span class="string">&#x27;path/to/pbmc68k.h5ad&#x27;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 加载数据集</span></span><br><span class="line">pbmc = sc.read(data_path)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 绘制UMAP图，并标记单个基因的表达</span></span><br><span class="line">sc.pl.umap(pbmc, color=<span class="string">&#x27;CD79A&#x27;</span>)</span><br><span class="line">plt.savefig(<span class="string">&#x27;plot/gene.pdf&#x27;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 绘制UMAP图，并标记多个基因的表达</span></span><br><span class="line">sc.pl.umap(pbmc, color=[<span class="string">&#x27;CD79A&#x27;</span>, <span class="string">&#x27;MS4A1&#x27;</span>, <span class="string">&#x27;GJ&#x27;</span>, <span class="string">&#x27;CD3D&#x27;</span>, <span class="string">&#x27;FCER1A&#x27;</span>, <span class="string">&#x27;FCGR3A&#x27;</span>, <span class="string">&#x27;n_counts&#x27;</span>, <span class="string">&quot;bulk labels&quot;</span>], s=<span class="number">50</span>, frameon=<span class="literal">False</span>, nCols=<span class="number">3</span>, vmax=<span class="string">&#x27;p99&#x27;</span>)</span><br><span class="line"><span class="comment">#s: 这是一个可选参数，用于指定细胞的标记大小。在这个例子中，我们将标记的大小设置为50。</span></span><br><span class="line"><span class="comment">#frameon: 这是一个布尔值，用于控制绘图的边框是否显示。在这个例子中，我们将其设置为False，即不显示边框。</span></span><br><span class="line"><span class="comment">#nCols: 这是一个整数，用于指定在绘图中每行显示的子图的列数。在这个例子中，我们将每行显示4个子图</span></span><br><span class="line"><span class="comment">#vmax: 这是一个可选参数，用于指定颜色映射的最大值。在这个例子中，我们将其设置为&#x27;p99&#x27;，表示使用数据的第99百分位作为颜色映射的最大值。</span></span><br><span class="line">plt.savefig(<span class="string">&#x27;plot/gene.pdf&#x27;</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/Cgf6barvXoxImrxUvqQcAv3Nnmd.png"></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> re</span><br><span class="line"><span class="keyword">from</span> matplotlib <span class="keyword">import</span> pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">from</span> matplotlib.pyplot <span class="keyword">import</span> rc_context</span><br><span class="line"></span><br><span class="line"><span class="comment"># 设置Scanpy的图像参数</span></span><br><span class="line">sc.set_figure_params(dpi=<span class="number">100</span>, color_map=<span class="string">&#x27;viridis_r&#x27;</span>)</span><br><span class="line">sc.settings.verbosity = <span class="number">1</span></span><br><span class="line">sc.logging.print_header()</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment"># 手动下载数据集</span></span><br><span class="line"><span class="comment">#绝对路径</span></span><br><span class="line">data_path = <span class="string">&#x27;path/to/pbmc68k.h5ad&#x27;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 加载数据集</span></span><br><span class="line">pbmc = sc.read(data_path)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 定义要标记的基因字典</span></span><br><span class="line"><span class="comment">#不同的细胞类型对应的细胞标签</span></span><br><span class="line">marker_genes_dict = &#123;</span><br><span class="line">    <span class="string">&#x27;B-cell&#x27;</span>: [<span class="string">&#x27;CD79A&#x27;</span>, <span class="string">&#x27;MS4A1&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Dendritic&#x27;</span>: [<span class="string">&#x27;FCER1A&#x27;</span>, <span class="string">&#x27;CST3&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Monocytes&#x27;</span>: [<span class="string">&#x27;FCGR3A&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;NK&#x27;</span>: [<span class="string">&#x27;GNLY&#x27;</span>, <span class="string">&#x27;NKG7&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Other&#x27;</span>: [<span class="string">&#x27;GLL1&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Plasma&#x27;</span>: [<span class="string">&#x27;IGJ&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;T-cell&#x27;</span>: [<span class="string">&#x27;CD3D&#x27;</span>]</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="comment"># 对细胞进行聚类</span></span><br><span class="line">sc.tl.leiden(pbmc, key_added=<span class="string">&#x27;clusters&#x27;</span>, resolution=<span class="number">0.5</span>)</span><br><span class="line"><span class="comment">#resolution：用于调节聚类的分辨率参数，值越大会产生更多的聚类簇，默认为1.0。</span></span><br><span class="line"><span class="comment">#n_neighbors：用于计算最近邻的邻居数量，默认为30。</span></span><br><span class="line"><span class="comment">#random_state：用于控制随机数生成器的种子，以确保可复现性。</span></span><br><span class="line"><span class="comment">#key_added：用于指定存储聚类结果的关键字，将聚类结果存储在adata.obs中，默认为leiden。</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 绘制基因的点图</span></span><br><span class="line">sc.pl.dotplot(pbmc, marker_genes_dict, groupby=<span class="string">&#x27;clusters&#x27;</span>, dendrogram=<span class="literal">True</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment"># 保存图像</span></span><br><span class="line">plt.savefig(<span class="string">&#x27;plot/gene.pdf&#x27;</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/DzYbbf9JNoX3Asxnet9cubYKnLe.png"></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"><span class="keyword">import</span> matplotlib.pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="comment"># 定义细胞簇到注释的映射字典</span></span><br><span class="line">cluster2annotation = &#123;</span><br><span class="line">    <span class="string">&#x27;0&#x27;</span>: <span class="string">&#x27;Monocytes&#x27;</span>,</span><br><span class="line">    <span class="string">&#x27;1&#x27;</span>: <span class="string">&#x27;Dendritic&#x27;</span>,</span><br><span class="line">    <span class="string">&#x27;2&#x27;</span>: <span class="string">&#x27;T-cell&#x27;</span>,</span><br><span class="line">    <span class="string">&#x27;3&#x27;</span>: <span class="string">&#x27;NK&#x27;</span>,</span><br><span class="line">    <span class="string">&#x27;4&#x27;</span>: <span class="string">&#x27;B-cell&#x27;</span>,</span><br><span class="line">    <span class="string">&#x27;5&#x27;</span>: <span class="string">&#x27;Dendritic&#x27;</span>,</span><br><span class="line">    <span class="string">&#x27;6&#x27;</span>: <span class="string">&#x27;Plasma&#x27;</span>,</span><br><span class="line">    <span class="string">&#x27;7&#x27;</span>: <span class="string">&#x27;Other&#x27;</span>,</span><br><span class="line">    <span class="string">&#x27;8&#x27;</span>: <span class="string">&#x27;Dendritic&#x27;</span></span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="comment"># 将注释添加到数据集</span></span><br><span class="line">pbmc.obs[<span class="string">&#x27;cell type&#x27;</span>] = pbmc.obs[<span class="string">&#x27;clusters&#x27;</span>].<span class="built_in">map</span>(cluster2annotation).astype(<span class="string">&#x27;category&#x27;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 绘制基因的点图</span></span><br><span class="line">sc.pl.dotplot(pbmc, marker_genes_dict, <span class="string">&#x27;cell type&#x27;</span>, dendrogram=<span class="literal">True</span>)</span><br><span class="line"><span class="comment">#显示树状图（dendrogram=True）</span></span><br><span class="line">plt.savefig(<span class="string">&#x27;plot/gene.pdf&#x27;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 绘制UMAP图并按细胞类型着色</span></span><br><span class="line">sc.pl.umap(pbmc, color=<span class="string">&#x27;cell type&#x27;</span>, legend_loc=<span class="string">&#x27;on data&#x27;</span>, frameon=<span class="literal">False</span>, legend_fontsize=<span class="number">10</span>, legend_fontoutline=<span class="number">2</span>)</span><br><span class="line"><span class="comment">#legend_loc=&#x27;on data&#x27;将图例放置在数据区域</span></span><br><span class="line">plt.savefig(<span class="string">&#x27;plot/umap.pdf&#x27;</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/DOIobQUbRon8fTxvTIQcZDjHnKg.png"></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> re</span><br><span class="line"><span class="keyword">from</span> matplotlib <span class="keyword">import</span> pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">from</span> matplotlib.pyplot <span class="keyword">import</span> rc_context</span><br><span class="line"></span><br><span class="line"><span class="comment"># 设置Scanpy的图像参数</span></span><br><span class="line">sc.set_figure_params(dpi=<span class="number">100</span>, color_map=<span class="string">&#x27;viridis_r&#x27;</span>)</span><br><span class="line">sc.settings.verbosity = <span class="number">1</span></span><br><span class="line">sc.logging.print_header()</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment"># 手动下载数据集</span></span><br><span class="line"><span class="comment">#绝对路径</span></span><br><span class="line">data_path = <span class="string">&#x27;path/to/pbmc68k.h5ad&#x27;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 加载数据集</span></span><br><span class="line">pbmc = sc.read(data_path)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 定义要标记的基因字典</span></span><br><span class="line"><span class="comment">#不同的细胞类型对应的细胞标签</span></span><br><span class="line">marker_genes_dict = &#123;</span><br><span class="line">    <span class="string">&#x27;B-cell&#x27;</span>: [<span class="string">&#x27;CD79A&#x27;</span>, <span class="string">&#x27;MS4A1&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Dendritic&#x27;</span>: [<span class="string">&#x27;FCER1A&#x27;</span>, <span class="string">&#x27;CST3&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Monocytes&#x27;</span>: [<span class="string">&#x27;FCGR3A&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;NK&#x27;</span>: [<span class="string">&#x27;GNLY&#x27;</span>, <span class="string">&#x27;NKG7&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Other&#x27;</span>: [<span class="string">&#x27;GLL1&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Plasma&#x27;</span>: [<span class="string">&#x27;IGJ&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;T-cell&#x27;</span>: [<span class="string">&#x27;CD3D&#x27;</span>]</span><br><span class="line">&#125;</span><br><span class="line"><span class="comment"># 绘制基因的小提琴图</span></span><br><span class="line">sc.pl.violin(pbmc, [<span class="string">&#x27;CD79A&#x27;</span>, <span class="string">&#x27;MS4A1&#x27;</span>], groupby=<span class="string">&#x27;clusters&#x27;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 保存小提琴图结果为PDF文件</span></span><br><span class="line">plt.savefig(<span class="string">&#x27;plot/gene.pdf&#x27;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 绘制基因的堆叠小提琴图</span></span><br><span class="line">ax = sc.pl.stacked_violin(pbmc, marker_genes_dict, groupby=<span class="string">&#x27;clusters&#x27;</span>, swap_axes=<span class="literal">False</span>, dendrogram=<span class="literal">True</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 保存堆叠小提琴图结果为PDF文件</span></span><br><span class="line">plt.savefig(<span class="string">&#x27;plot/gene_stacked_violin.pdf&#x27;</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/Fd6eb7k4Bo0b32xOeTGcmW9wnqg.png"></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> re</span><br><span class="line"><span class="keyword">from</span> matplotlib <span class="keyword">import</span> pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">from</span> matplotlib.pyplot <span class="keyword">import</span> rc_context</span><br><span class="line"></span><br><span class="line"><span class="comment"># 设置Scanpy的图像参数</span></span><br><span class="line">sc.set_figure_params(dpi=<span class="number">100</span>, color_map=<span class="string">&#x27;viridis_r&#x27;</span>)</span><br><span class="line">sc.settings.verbosity = <span class="number">1</span></span><br><span class="line">sc.logging.print_header()</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment"># 手动下载数据集</span></span><br><span class="line"><span class="comment">#绝对路径</span></span><br><span class="line">data_path = <span class="string">&#x27;path/to/pbmc68k.h5ad&#x27;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 加载数据集</span></span><br><span class="line">pbmc = sc.read(data_path)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 定义要标记的基因字典</span></span><br><span class="line"><span class="comment">#不同的细胞类型对应的细胞标签</span></span><br><span class="line">marker_genes_dict = &#123;</span><br><span class="line">    <span class="string">&#x27;B-cell&#x27;</span>: [<span class="string">&#x27;CD79A&#x27;</span>, <span class="string">&#x27;MS4A1&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Dendritic&#x27;</span>: [<span class="string">&#x27;FCER1A&#x27;</span>, <span class="string">&#x27;CST3&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Monocytes&#x27;</span>: [<span class="string">&#x27;FCGR3A&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;NK&#x27;</span>: [<span class="string">&#x27;GNLY&#x27;</span>, <span class="string">&#x27;NKG7&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Other&#x27;</span>: [<span class="string">&#x27;GLL1&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Plasma&#x27;</span>: [<span class="string">&#x27;IGJ&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;T-cell&#x27;</span>: [<span class="string">&#x27;CD3D&#x27;</span>]</span><br><span class="line">&#125;</span><br><span class="line">sc.pl.matrixplot(pbmc, marker_genes_dict, <span class="string">&#x27;clusters&#x27;</span>, dendrogram=<span class="literal">True</span>, cmap=<span class="string">&#x27;Blues&#x27;</span>, standard_scale=<span class="string">&#x27;var&#x27;</span>, colorbar_title=<span class="string">&#x27;column scaled expression&#x27;</span>)</span><br><span class="line"><span class="comment">#通过设置standard_scale=&#x27;var&#x27;对列进行标准化处理，使得每列的表达量在同一尺度上。</span></span><br><span class="line">plt.savefig(<span class="string">&#x27;plot/gene.pdf&#x27;</span>)</span><br><span class="line"></span><br><span class="line">pbmc.layers[<span class="string">&#x27;scaled&#x27;</span>] = sc.pp.scale(pbmc, copy=<span class="literal">True</span>).X</span><br><span class="line"><span class="comment">#第二段代码绘制了另一个矩阵图，但这次使用的是经过缩放处理后的数据层。</span></span><br><span class="line"><span class="comment">#首先，通过sc.pp.scale函数对数据进行标准化处理，并将结果存储在数据集对象的layers属性中的scaled层。</span></span><br><span class="line">sc.pl.matrixplot(pbmc, marker_genes_dict, <span class="string">&#x27;clusters&#x27;</span>, dendrogram=<span class="literal">True</span>, colorbar_title=<span class="string">&#x27;mean z-score&#x27;</span>, layer=<span class="string">&#x27;scaled&#x27;</span>, vmin=-<span class="number">2</span>, vmax=<span class="number">2</span>, cmap=<span class="string">&#x27;RdBu_r&#x27;</span>)</span><br><span class="line"><span class="comment">#使用layer=&#x27;scaled&#x27;指定要绘制的数据层为scaled层。</span></span><br><span class="line"><span class="comment">#通过设置vmin=-2和vmax=2，可以限定颜色映射的范围在-2到2之间，以突出显示较小和较大的表达差异。</span></span><br><span class="line">plt.savefig(<span class="string">&#x27;plot/gene.pdf&#x27;</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/PCg4bJapmoKQr6xIBmXcYmeFn8f.png"></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> re</span><br><span class="line"><span class="keyword">from</span> matplotlib <span class="keyword">import</span> pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">from</span> matplotlib.pyplot <span class="keyword">import</span> rc_context</span><br><span class="line"></span><br><span class="line"><span class="comment"># 设置Scanpy的图像参数</span></span><br><span class="line">sc.set_figure_params(dpi=<span class="number">100</span>, color_map=<span class="string">&#x27;viridis_r&#x27;</span>)</span><br><span class="line">sc.settings.verbosity = <span class="number">1</span></span><br><span class="line">sc.logging.print_header()</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment"># 手动下载数据集</span></span><br><span class="line"><span class="comment">#绝对路径</span></span><br><span class="line">data_path = <span class="string">&#x27;path/to/pbmc68k.h5ad&#x27;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 加载数据集</span></span><br><span class="line">pbmc = sc.read(data_path)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 定义要标记的基因字典</span></span><br><span class="line"><span class="comment">#不同的细胞类型对应的细胞标签</span></span><br><span class="line">marker_genes_dict = &#123;</span><br><span class="line">    <span class="string">&#x27;B-cell&#x27;</span>: [<span class="string">&#x27;CD79A&#x27;</span>, <span class="string">&#x27;MS4A1&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Dendritic&#x27;</span>: [<span class="string">&#x27;FCER1A&#x27;</span>, <span class="string">&#x27;CST3&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Monocytes&#x27;</span>: [<span class="string">&#x27;FCGR3A&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;NK&#x27;</span>: [<span class="string">&#x27;GNLY&#x27;</span>, <span class="string">&#x27;NKG7&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Other&#x27;</span>: [<span class="string">&#x27;GLL1&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Plasma&#x27;</span>: [<span class="string">&#x27;IGJ&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;T-cell&#x27;</span>: [<span class="string">&#x27;CD3D&#x27;</span>]</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">ax = sc.pl.heatmap(pbmc, marker_genes_dict, groupby=<span class="string">&#x27;clusters&#x27;</span>, cmap=<span class="string">&#x27;viridis&#x27;</span>, dendrogram=<span class="literal">True</span>)</span><br><span class="line">plt.savefig(<span class="string">&#x27;plot/gene.pdf&#x27;</span>)</span><br><span class="line"></span><br><span class="line">pbmc.layers[<span class="string">&#x27;scaled&#x27;</span>] = sc.pp.scale(pbmc, copy=<span class="literal">True</span>).X</span><br><span class="line">ax = sc.pl.heatmap(pbmc, marker_genes_dict, groupby=<span class="string">&#x27;clusters&#x27;</span>, layer=<span class="string">&#x27;scaled&#x27;</span>, vmin=-<span class="number">2</span>, vmax=<span class="number">2</span>, cmap=<span class="string">&#x27;RdBu_r&#x27;</span>, dendrogram=<span class="literal">True</span>, swap_axes=<span class="literal">True</span>, figsize=(<span class="number">11</span>,<span class="number">4</span>))</span><br><span class="line"><span class="comment">#通过设置swap_axes=True，交换热图的行和列，使得基因显示在纵向，细胞簇显示在横向。</span></span><br><span class="line"><span class="comment">#还设置了figsize=(11,4)来调整图像的大小。</span></span><br><span class="line">plt.savefig(<span class="string">&#x27;plot/gene.pdf&#x27;</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/FrsrbabJgoixhBxLolNczJI8nhe.png"></p>
<p><img src="/post/single-cell/DyZAbbLGaoN3axxUsP0cpqKyncc.png"></p>
<p><strong>踪迹图</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> re</span><br><span class="line"><span class="keyword">from</span> matplotlib <span class="keyword">import</span> pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">from</span> matplotlib.pyplot <span class="keyword">import</span> rc_context</span><br><span class="line"></span><br><span class="line"><span class="comment"># 设置Scanpy的图像参数</span></span><br><span class="line">sc.set_figure_params(dpi=<span class="number">100</span>, color_map=<span class="string">&#x27;viridis_r&#x27;</span>)</span><br><span class="line">sc.settings.verbosity = <span class="number">1</span></span><br><span class="line">sc.logging.print_header()</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment"># 手动下载数据集</span></span><br><span class="line"><span class="comment">#绝对路径</span></span><br><span class="line">data_path = <span class="string">&#x27;path/to/pbmc68k.h5ad&#x27;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 加载数据集</span></span><br><span class="line">pbmc = sc.read(data_path)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 定义要标记的基因字典</span></span><br><span class="line"><span class="comment">#不同的细胞类型对应的细胞标签</span></span><br><span class="line">marker_genes_dict = &#123;</span><br><span class="line">    <span class="string">&#x27;B-cell&#x27;</span>: [<span class="string">&#x27;CD79A&#x27;</span>, <span class="string">&#x27;MS4A1&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Dendritic&#x27;</span>: [<span class="string">&#x27;FCER1A&#x27;</span>, <span class="string">&#x27;CST3&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Monocytes&#x27;</span>: [<span class="string">&#x27;FCGR3A&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;NK&#x27;</span>: [<span class="string">&#x27;GNLY&#x27;</span>, <span class="string">&#x27;NKG7&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Other&#x27;</span>: [<span class="string">&#x27;GLL1&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Plasma&#x27;</span>: [<span class="string">&#x27;IGJ&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;T-cell&#x27;</span>: [<span class="string">&#x27;CD3D&#x27;</span>]</span><br><span class="line">&#125;</span><br><span class="line">ax = sc.pl.tracksplot(pbmc, marker_genes_dict, groupby=<span class="string">&#x27;clusters&#x27;</span>, dendrogram=<span class="literal">True</span>)</span><br><span class="line"><span class="comment">#vmax：设置颜色映射的最大值，用于控制颜色的亮度，默认为 None，即自动确定最大值。</span></span><br><span class="line"><span class="comment">#vmin：设置颜色映射的最小值，用于控制颜色的亮度，默认为 None，即自动确定最小值。</span></span><br><span class="line">plt.savefig(<span class="string">&#x27;plot/gene.pdf&#x27;</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/ClkobqAh2o6ZEDxWZ7rchQexnsb.png"></p>
<h2 id="差异基因表达"><a href="#差异基因表达" class="headerlink" title="差异基因表达"></a>差异基因表达</h2><blockquote>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br></pre></td><td class="code"><pre><span class="line">```</span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> re</span><br><span class="line"><span class="keyword">from</span> matplotlib <span class="keyword">import</span> pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">from</span> matplotlib.pyplot <span class="keyword">import</span> rc_context</span><br><span class="line"></span><br><span class="line"><span class="comment"># 设置 Scanpy 的图像参数</span></span><br><span class="line"></span><br><span class="line">sc.set_figure_params(dpi=<span class="number">100</span>, color_map=<span class="string">&#x27;viridis_r&#x27;</span>)</span><br><span class="line">sc.settings.verbosity = <span class="number">1</span></span><br><span class="line">sc.logging.print_header()</span><br><span class="line"></span><br><span class="line"><span class="comment"># 手动下载数据集</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#绝对路径</span></span><br><span class="line">data_path = <span class="string">&#x27;path/to/pbmc68k.h5ad&#x27;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 加载数据集</span></span><br><span class="line"></span><br><span class="line">pbmc = sc.read(data_path)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 定义要标记的基因字典</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#不同的细胞类型对应的细胞标签</span></span><br><span class="line">marker_genes_dict = &#123;</span><br><span class="line"><span class="string">&#x27;B-cell&#x27;</span>: [<span class="string">&#x27;CD79A&#x27;</span>, <span class="string">&#x27;MS4A1&#x27;</span>],</span><br><span class="line"><span class="string">&#x27;Dendritic&#x27;</span>: [<span class="string">&#x27;FCER1A&#x27;</span>, <span class="string">&#x27;CST3&#x27;</span>],</span><br><span class="line"><span class="string">&#x27;Monocytes&#x27;</span>: [<span class="string">&#x27;FCGR3A&#x27;</span>],</span><br><span class="line"><span class="string">&#x27;NK&#x27;</span>: [<span class="string">&#x27;GNLY&#x27;</span>, <span class="string">&#x27;NKG7&#x27;</span>],</span><br><span class="line"><span class="string">&#x27;Other&#x27;</span>: [<span class="string">&#x27;GLL1&#x27;</span>],</span><br><span class="line"><span class="string">&#x27;Plasma&#x27;</span>: [<span class="string">&#x27;IGJ&#x27;</span>],</span><br><span class="line"><span class="string">&#x27;T-cell&#x27;</span>: [<span class="string">&#x27;CD3D&#x27;</span>]</span><br><span class="line">&#125;</span><br><span class="line">sc.tl.rank_genes_groups(pbmc, groupby=<span class="string">&#x27;clusters&#x27;</span>, method=<span class="string">&#x27;wilcoxon&#x27;</span>)</span><br><span class="line">sc.pl.rank_genes_groups_dotplot(pbmc, n_genes=<span class="number">4</span>)</span><br><span class="line">plt.savefig(<span class="string">&#x27;plot/gene.pdf&#x27;</span>)</span><br><span class="line"><span class="comment">#通过设置 use_raw=False，函数将使用差异表达数据进行绘制，</span></span><br><span class="line"><span class="comment">#其中 &quot;mean expression in group&quot; 将被替换为 &quot;log fold change&quot;。</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>
</blockquote>
<p><img src="/post/single-cell/TgZIbiA1vozqsex6uaRcLzdkn0f.png"></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> re</span><br><span class="line"><span class="keyword">from</span> matplotlib <span class="keyword">import</span> pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">from</span> matplotlib.pyplot <span class="keyword">import</span> rc_context</span><br><span class="line"></span><br><span class="line"><span class="comment"># 设置Scanpy的图像参数</span></span><br><span class="line">sc.set_figure_params(dpi=<span class="number">100</span>, color_map=<span class="string">&#x27;viridis_r&#x27;</span>)</span><br><span class="line">sc.settings.verbosity = <span class="number">1</span></span><br><span class="line">sc.logging.print_header()</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment"># 手动下载数据集</span></span><br><span class="line"><span class="comment">#绝对路径</span></span><br><span class="line">data_path = <span class="string">&#x27;path/to/pbmc68k.h5ad&#x27;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 加载数据集</span></span><br><span class="line">pbmc = sc.read(data_path)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 定义要标记的基因字典</span></span><br><span class="line"><span class="comment">#不同的细胞类型对应的细胞标签</span></span><br><span class="line">marker_genes_dict = &#123;</span><br><span class="line">    <span class="string">&#x27;B-cell&#x27;</span>: [<span class="string">&#x27;CD79A&#x27;</span>, <span class="string">&#x27;MS4A1&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Dendritic&#x27;</span>: [<span class="string">&#x27;FCER1A&#x27;</span>, <span class="string">&#x27;CST3&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Monocytes&#x27;</span>: [<span class="string">&#x27;FCGR3A&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;NK&#x27;</span>: [<span class="string">&#x27;GNLY&#x27;</span>, <span class="string">&#x27;NKG7&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Other&#x27;</span>: [<span class="string">&#x27;GLL1&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;Plasma&#x27;</span>: [<span class="string">&#x27;IGJ&#x27;</span>],</span><br><span class="line">    <span class="string">&#x27;T-cell&#x27;</span>: [<span class="string">&#x27;CD3D&#x27;</span>]</span><br><span class="line">&#125;</span><br><span class="line"><span class="keyword">with</span> rc_context(&#123;<span class="string">&#x27;figure.figsize&#x27;</span>: (<span class="number">9</span>, <span class="number">1.5</span>)&#125;):</span><br><span class="line">    sc.pl.rank_genes_groups_violin(pbmc, n_genes=<span class="number">20</span>, groups=<span class="string">&#x27;0&#x27;</span>, jitter=<span class="literal">False</span>)</span><br><span class="line">plt.savefig(<span class="string">&#x27;plot/gene.pdf&#x27;</span>)</span><br><span class="line"><span class="comment">#这段代码使用 rc_context 上下文管理器来设置绘图参数，将图像的宽度设置为9英寸，高度设置为1.5英寸。</span></span><br><span class="line">sc.pl.rank_genes_groups_violin(</span><br><span class="line">    adata,             <span class="comment"># AnnData object</span></span><br><span class="line">    groupby=<span class="literal">None</span>,      <span class="comment"># 列名，用于指定细胞群组的分类变量</span></span><br><span class="line">    use_raw=<span class="literal">True</span>,      <span class="comment"># 是否使用原始数据进行绘图，默认为 True</span></span><br><span class="line">    log=<span class="literal">False</span>,         <span class="comment"># 是否对表达值进行对数转换，默认为 False</span></span><br><span class="line">    stripplot=<span class="literal">True</span>,    <span class="comment"># 是否在小提琴图上显示散点图，默认为 True</span></span><br><span class="line">    jitter=<span class="literal">True</span>,       <span class="comment"># 是否对散点图进行抖动处理，默认为 True</span></span><br><span class="line">    size=<span class="number">1</span>,            <span class="comment"># 散点图的点的大小，默认为 1</span></span><br><span class="line">    layer=<span class="literal">None</span>,        <span class="comment"># 如果数据有多个层，可以指定要使用的层，默认为 None</span></span><br><span class="line">    show=<span class="literal">False</span>,        <span class="comment"># 是否显示绘图，默认为 False</span></span><br><span class="line">    save=<span class="literal">None</span>,         <span class="comment"># 图像保存路径，默认为 None</span></span><br><span class="line">    ax=<span class="literal">None</span>,           <span class="comment"># 绘图所使用的轴对象，默认为 None</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/QhkMbW3cEo15Q8xsq2ecKkQ2n3b.png"></p>
<h2 id="聚类树"><a href="#聚类树" class="headerlink" title="聚类树"></a>聚类树</h2><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> re</span><br><span class="line"><span class="keyword">from</span> matplotlib <span class="keyword">import</span> pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">from</span> matplotlib.pyplot <span class="keyword">import</span> rc_context</span><br><span class="line"></span><br><span class="line"><span class="comment"># 设置Scanpy的图像参数</span></span><br><span class="line">sc.set_figure_params(dpi=<span class="number">100</span>, color_map=<span class="string">&#x27;viridis_r&#x27;</span>)</span><br><span class="line">sc.settings.verbosity = <span class="number">1</span></span><br><span class="line">sc.logging.print_header()</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment"># 手动下载数据集</span></span><br><span class="line"><span class="comment">#绝对路径</span></span><br><span class="line">data_path = <span class="string">&#x27;path/to/pbmc68k.h5ad&#x27;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 加载数据集</span></span><br><span class="line">pbmc = sc.read(data_path)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">sc.tl.dendrogram(pbmc, <span class="string">&quot;bulk_labels&quot;</span>)  <span class="comment"># 计算层次聚类的树状图,基于细胞标签</span></span><br><span class="line">ax = sc.pl.dendrogram(pbmc, <span class="string">&quot;bulk_labels&quot;</span>)  <span class="comment"># 绘制树状图，并返回绘图所使用的轴对象</span></span><br><span class="line">plt.savefig(<span class="string">&quot;plot/gene.pdf&quot;</span>)  <span class="comment"># 保存绘图结果到文件</span></span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/MQjdbuR8robBqpxI30JcShb8n5e.png"></p>
<h2 id="相关性"><a href="#相关性" class="headerlink" title="相关性"></a>相关性</h2><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> re</span><br><span class="line"><span class="keyword">from</span> matplotlib <span class="keyword">import</span> pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">from</span> matplotlib.pyplot <span class="keyword">import</span> rc_context</span><br><span class="line"></span><br><span class="line"><span class="comment"># 设置Scanpy的图像参数</span></span><br><span class="line">sc.set_figure_params(dpi=<span class="number">100</span>, color_map=<span class="string">&#x27;viridis_r&#x27;</span>)</span><br><span class="line">sc.settings.verbosity = <span class="number">1</span></span><br><span class="line">sc.logging.print_header()</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment"># 手动下载数据集</span></span><br><span class="line"><span class="comment">#绝对路径</span></span><br><span class="line">data_path = <span class="string">&#x27;path/to/pbmc68k.h5ad&#x27;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 加载数据集</span></span><br><span class="line">pbmc = sc.read(data_path)</span><br><span class="line">ax = sc.pl.correlation_matrix(pbmc, <span class="string">&quot;bulk_labels&quot;</span>, figsize=(<span class="number">5</span>, <span class="number">3.5</span>))  <span class="comment"># 绘制相关性矩阵，并返回绘图所使用的轴对象</span></span><br><span class="line">plt.savefig(<span class="string">&quot;plot/gene.pdf&quot;</span>)  <span class="comment"># 保存绘图结果到文件</span></span><br><span class="line"><span class="comment">#请注意，示例中的 &quot;bulk_labels&quot; 应该是您数据集中真实存在的变量名，您需要根据您的数据集进行相应的修改。</span></span><br><span class="line"><span class="comment">#另外，您也可以根据需要调整 figsize 参数来设置绘图的大小。</span></span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/PxWvbJMSqocvpTxxeITcEdA7nXg.png"></p>
<h1 id="实操"><a href="#实操" class="headerlink" title="实操"></a>实操</h1><p><img src="/post/single-cell/LsxpbAIBkoNp1Kx3fq8cj8REnQe.png"></p>
<h1 id="umap"><a href="#umap" class="headerlink" title="umap"></a>umap</h1><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"><span class="keyword">from</span> matplotlib <span class="keyword">import</span> pyplot <span class="keyword">as</span> plt</span><br><span class="line">sc.tl.umap(adata)</span><br><span class="line">sc.pl.umap(adata,color=<span class="string">&#x27;leiden&#x27;</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/post/single-cell/KAnMb9tk4oYQ2Uxz1Vgc0pa0n60.png"></p>
<p><img src="/post/single-cell/Q3gYb0GTiopuOGxQyWWcgJ9GnTe.png"></p>
<p>较小的 pval_adj 值表示差异更显著。</p>
<h1 id="批次处理"><a href="#批次处理" class="headerlink" title="批次处理"></a>批次处理</h1><p>批次效应是测量的表达水平的变化，这是处理不同组或“批次”中的细胞的结果。例如，如果两个实验室从同一队列中采集样本，但这些样本的解离方式不同，则可能会出现批次效应。如果实验室 A 优化其解离方案以解离样品中的细胞，同时最大限度地减少对细胞的压力，而实验室 B 没有这样做，那么 B 组数据中的细胞很可能会表达更多与压力相关的基因</p>
<p><strong>同一批次不需要进行批次处理，不同批次才需要，同一批次批次效应比较小。</strong></p>
<p><strong>批次效应在行的质控之后。对列的处理。</strong></p>
<p><img src="/post/single-cell/IfL2buYF5oaaRrxFdQ7c3B3Nnfg.png"></p>
<p><img src="/post/single-cell/IVEGbhipQoAbHPx6C6McCar8noh.png"></p>
<p><img src="/post/single-cell/C4qzb50szoUTuJxnVEYcs4X7nNg.png"></p>
<p><img src="/post/single-cell/QpXhbKYnXootWHxENTRcn1OonLb.png"></p>
<p>绝大多数的批次效应校正算法都包括了如何构建一个稳定的数学模型。我们按照模型的发展顺序，可以分为四类：<strong>全局模型</strong>，<strong>线性嵌入模型</strong>，<strong>基于图的模型</strong>，<strong>深度学习****模型</strong></p>
<ul>
<li>全局模型: 源自 bulk RNA-seq，将批次效应建模为所有细胞中存在的（加法&#x2F;或乘法）效应。一个常见的例子是 ComBat</li>
<li>线性嵌入模型: 是第一个单细胞特异性批量去除方法。这些方法通常使用奇异值分解 (SVD) 的变体来嵌入数据，然后在嵌入中跨批次查找相似单元的局部邻域，并使用它们以局部自适应（非线性）方式校正批次效应。常见的例子包括最近邻 MNN，Seurat，Harmony，Scanorama，FastMNN 等</li>
<li>基于图的模型: 通常是运行速度最快的方法。使用最近邻域图来表示每个批次的数据。通过强制连接不同批次的细胞，然后修建细胞类型组成的差异的图的边缘，可以纠正批次效应。一个常见的例子是 BBKNN</li>
<li>深度学习模型: 大多数深度学习批次效应校正方法都基于自动编码器网络，并且要么在条件变分自动编码器（CVAE）中对批量协变量进行降维，要么在嵌入空间中拟合局部线性校正。</li>
</ul>
</article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta"><i class="fas fa-circle-user fa-fw"></i>文章作者: </span><span class="post-copyright-info"><a href="https://glking123.github.io">Lei Lei</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta"><i class="fas fa-square-arrow-up-right fa-fw"></i>文章链接: </span><span class="post-copyright-info"><a href="https://glking123.github.io/2024/06/02/single-cell/">https://glking123.github.io/2024/06/02/single-cell/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta"><i class="fas fa-circle-exclamation fa-fw"></i>版权声明: </span><span class="post-copyright-info">本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank">CC BY-NC-SA 4.0</a> 许可协议。转载请注明来自 <a href="https://glking123.github.io" target="_blank">LeiLeiBear</a>！</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF/">生物信息</a></div><div class="post_share"><div class="social-share" data-image="/img/shengwu.jpg" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1.1.3/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1.1.3/sharejs/dist/js/social-share.min.js" defer></script></div></div><div class="post-reward"><div class="reward-button"><i class="fas fa-qrcode"></i>赞助</div><div class="reward-main"><ul class="reward-all"><li class="reward-item"><a href="/img/wechat.jpg" target="_blank"><img class="post-qr-code-img" src="/img/wechat.jpg" alt="微信"/></a><div class="post-qr-code-desc">微信</div></li><li class="reward-item"><a href="/img/alipay.jpg" target="_blank"><img class="post-qr-code-img" src="/img/alipay.jpg" alt="支付宝"/></a><div class="post-qr-code-desc">支付宝</div></li></ul></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2024/06/30/Array/" title="Array"><img class="cover" src="/img/jisuanji.jpg" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">上一篇</div><div class="prev_info">Array</div></div></a></div><div class="next-post pull-right"><a href="/2024/05/31/hello-world/" title="Hello World"><div class="cover" style="background: var(--default-bg-color)"></div><div class="pagination-info"><div class="label">下一篇</div><div class="next_info">Hello World</div></div></a></div></nav></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="is-center"><div class="avatar-img"><img src="/img/touxiang.jpg" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="author-info__name">Lei Lei</div><div class="author-info__description">Never Give Up</div></div><div class="card-info-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">5</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">2</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">2</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/GLking123"><i class="fab fa-github"></i><span>Follow Me</span></a></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>公告</span></div><div class="announcement_content">希望可以对你有所帮助😊</div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>目录</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%8D%95%E7%BB%86%E8%83%9E"><span class="toc-number">1.</span> <span class="toc-text">单细胞</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%84%E5%92%8C%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%E7%9A%84%E5%8C%BA%E5%88%AB%E5%92%8C%E9%80%89%E6%8B%A9"><span class="toc-number">2.</span> <span class="toc-text">单细胞转录组和转录组分析的区别和选择</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E7%A9%BA%E9%97%B4%E8%BD%AC%E5%BD%95%E7%BB%84%E5%92%8C%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%84%E5%8C%BA%E5%88%AB%E5%92%8C%E9%80%89%E6%8B%A9"><span class="toc-number">3.</span> <span class="toc-text">空间转录组和单细胞转录组区别和选择</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#python-%E7%89%88"><span class="toc-number">4.</span> <span class="toc-text">python 版</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#1%E3%80%81%E5%8D%95%E7%BB%86%E8%83%9E%E6%B5%8B%E5%BA%8F%E5%8F%91%E5%B1%95%E5%8E%86%E7%A8%8B"><span class="toc-number">5.</span> <span class="toc-text">1、单细胞测序发展历程</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#2%E3%80%81%E7%9B%AE%E5%89%8D%E5%8D%95%E7%BB%86%E8%83%9E%E6%B5%8B%E5%BA%8F%E7%B1%BB%E5%9E%8B"><span class="toc-number">6.</span> <span class="toc-text">2、目前单细胞测序类型</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#10X-%E5%8D%95%E7%BB%86%E8%83%9E%E6%B5%8B%E5%BA%8F%E6%96%87%E4%BB%B6"><span class="toc-number">6.1.</span> <span class="toc-text">*10X 单细胞测序文件*</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%84%E6%95%B0%E6%8D%AE%E8%B4%A8%E6%8E%A7%EF%BC%88QC%EF%BC%89%E5%8F%8A%E5%90%88%E5%B9%B6%E5%8E%BB%E9%99%A4%E6%89%B9%E6%AC%A1%E6%95%88%E5%BA%94"><span class="toc-number">7.</span> <span class="toc-text">单细胞转录组数据质控（QC）及合并去除批次效应</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#cellranger-%E5%8D%95%E7%BB%86%E8%83%9E"><span class="toc-number">8.</span> <span class="toc-text">cellranger 单细胞</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%BB%93%E6%9E%9C%E5%88%86%E6%9E%90"><span class="toc-number">8.1.</span> <span class="toc-text">结果分析</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%88%9B%E5%BB%BA-anndata-%E5%AF%B9%E8%B1%A1-%E5%92%8C-QC"><span class="toc-number">9.</span> <span class="toc-text">创建 anndata 对象 和 QC</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E9%A2%84%E5%A4%84%E7%90%86"><span class="toc-number">9.1.</span> <span class="toc-text">**预处理 **</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%98%BE%E7%A4%BA%E5%9C%A8%E6%89%80%E6%9C%89%E7%BB%86%E8%83%9E%E4%B8%AD%E5%9C%A8%E6%AF%8F%E4%B8%AA%E5%8D%95%E7%BB%86%E8%83%9E%E4%B8%AD%E4%BA%A7%E7%94%9F%E6%9C%80%E9%AB%98%E8%AE%A1%E6%95%B0%E5%88%86%E6%95%B0%E7%9A%84%E5%9F%BA%E5%9B%A0"><span class="toc-number">9.1.1.</span> <span class="toc-text">显示在所有细胞中在每个单细胞中产生最高计数分数的基因</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%BF%87%E6%BB%A4%E4%BD%8E%E8%B4%A8%E9%87%8F%E7%BB%86%E8%83%9E%E6%A0%B7%E6%9C%AC-pp-preprocessing"><span class="toc-number">9.1.2.</span> <span class="toc-text">过滤低质量细胞样本 pp preprocessing</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%BF%87%E6%BB%A4%E5%8C%85%E5%90%AB%E7%BA%BF%E7%B2%92%E4%BD%93%E5%9F%BA%E5%9B%A0%E5%92%8C%E8%A1%A8%E8%BE%BE%E5%9F%BA%E5%9B%A0%E8%BF%87%E5%A4%9A%E7%9A%84%E7%BB%86%E8%83%9E"><span class="toc-number">9.1.3.</span> <span class="toc-text">过滤包含线粒体基因和表达基因过多的细胞</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%95%B0%E6%8D%AE%E7%9A%84%E6%A0%87%E5%87%86%E5%8C%96-%E5%BD%92%E4%B8%80%E5%8C%96"><span class="toc-number">9.2.</span> <span class="toc-text">**数据的标准化&#x2F;**归一化</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%A0%87%E5%87%86%E5%8C%96"><span class="toc-number">9.2.1.</span> <span class="toc-text">标准化</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%B0%86-AnnData-%E5%AF%B9%E8%B1%A1%E7%9A%84-%E3%80%82raw-%E5%B1%9E%E6%80%A7%E8%AE%BE%E7%BD%AE%E4%B8%BA%E5%BD%92%E4%B8%80%E5%8C%96%E5%92%8C%E5%AF%B9%E6%95%B0%E5%8C%96%E7%9A%84%E5%8E%9F%E5%A7%8B%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%EF%BC%88%E5%AD%98%E5%82%A8%E6%95%B0%E6%8D%AE%EF%BC%89"><span class="toc-number">9.2.2.</span> <span class="toc-text">将 AnnData 对象的 。raw 属性设置为归一化和对数化的原始基因表达（存储数据）</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%8E%B7%E5%8F%96%E5%8F%AA%E6%9C%89%E7%89%B9%E5%BC%82%E6%80%A7%E5%9F%BA%E5%9B%A0%E7%9A%84%E6%95%B0%E6%8D%AE%E9%9B%86"><span class="toc-number">9.2.3.</span> <span class="toc-text">获取只有特异性基因的数据集</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%A6%82%E4%BD%95%E8%AF%84%E4%BC%B0%E7%BB%86%E8%83%9E%E5%91%A8%E6%9C%9F%E7%9A%84%E5%BD%B1%E5%93%8D%EF%BC%9F"><span class="toc-number">9.2.4.</span> <span class="toc-text">如何评估细胞周期的影响？</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E4%B8%BB%E6%88%90%E5%88%86%E5%88%86%E6%9E%90-%EF%BC%88%E6%95%B0%E6%8D%AE%E9%99%8D%E7%BB%B4%EF%BC%89"><span class="toc-number">9.3.</span> <span class="toc-text">主成分分析****（数据降维）</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E9%80%9A%E8%BF%87%E8%BF%90%E8%A1%8C%E4%B8%BB%E6%88%90%E5%88%86%E5%88%86%E6%9E%90-%EF%BC%88PCA%EF%BC%89-%E6%9D%A5%E9%99%8D%E4%BD%8E%E6%95%B0%E6%8D%AE%E7%9A%84%E7%BB%B4%E6%95%B0%EF%BC%8C%E5%8F%AF%E4%BB%A5%E5%AF%B9%E6%95%B0%E6%8D%AE%E8%BF%9B%E8%A1%8C%E5%8E%BB%E5%99%AA%E5%B9%B6%E6%8F%AD%E7%A4%BA%E4%B8%8D%E5%90%8C%E5%88%86%E7%BE%A4%E7%9A%84%E4%B8%BB%E5%9B%A0%E7%B4%A0%E3%80%82"><span class="toc-number">9.3.1.</span> <span class="toc-text">通过运行主成分分析 （PCA） 来降低数据的维数，可以对数据进行去噪并揭示不同分群的主因素。</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%A3%80%E6%9F%A5%E5%8D%95%E4%B8%AA-PC-%E5%AF%B9%E6%95%B0%E6%8D%AE%E6%80%BB%E6%96%B9%E5%B7%AE%E7%9A%84%E8%B4%A1%E7%8C%AE%EF%BC%8C%E8%BF%99%E5%8F%AF%E4%BB%A5%E6%8F%90%E4%BE%9B%E7%BB%99%E6%88%91%E4%BB%AC%E5%BA%94%E8%AF%A5%E8%80%83%E8%99%91%E5%A4%9A%E5%B0%91%E4%B8%AA-PC-%E4%BB%A5%E8%AE%A1%E7%AE%97%E7%BB%86%E8%83%9E%E7%9A%84%E9%82%BB%E5%9F%9F%E5%85%B3%E7%B3%BB%E7%9A%84%E4%BF%A1%E6%81%AF%EF%BC%8C"><span class="toc-number">9.3.2.</span> <span class="toc-text">检查单个 PC 对数据总方差的贡献，这可以提供给我们应该考虑多少个 PC 以计算细胞的邻域关系的信息，</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%BB%86%E8%83%9E%E5%88%86%E7%BE%A4"><span class="toc-number">9.4.</span> <span class="toc-text">细胞分群</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%A3%80%E7%B4%A2%E7%94%9F%E7%89%A9%E5%AD%A6%E6%A0%87%E8%AE%B0%E5%9F%BA%E5%9B%A0"><span class="toc-number">9.5.</span> <span class="toc-text">检索生物学标记基因</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#T-test"><span class="toc-number">9.5.1.</span> <span class="toc-text">T-test</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Wilcoxon-rank-sum"><span class="toc-number">9.5.2.</span> <span class="toc-text">Wilcoxon rank-sum</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E9%80%BB%E8%BE%91%E5%9B%9E%E5%BD%92"><span class="toc-number">9.5.3.</span> <span class="toc-text">逻辑回归</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%BD%BD%E5%85%A5%E6%95%B0%E6%8D%AE"><span class="toc-number">9.5.4.</span> <span class="toc-text">载入数据</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%8E%B7%E5%8F%96%E8%81%9A%E7%B1%BB%E5%88%86%E7%BB%84%E5%92%8C%E5%88%86%E6%95%B0"><span class="toc-number">9.5.5.</span> <span class="toc-text">获取聚类分组和分数</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E9%87%8D%E8%A6%81%E6%96%B9%E6%B3%95"><span class="toc-number">9.6.</span> <span class="toc-text">重要方法</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%89%BE%E5%88%B0%E7%89%B9%E5%AE%9A%E7%9A%84%E5%9F%BA%E5%9B%A0-zika"><span class="toc-number">9.7.</span> <span class="toc-text">找到特定的基因-zika</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#Group-1-%E4%B8%8E-Group-2-%E6%AF%94%E8%BE%83%EF%BC%8C%E8%BF%9B%E8%A1%8C%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90"><span class="toc-number">9.7.1.</span> <span class="toc-text">Group 1 与 Group 2 比较，进行差异分析</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Group-0-%E4%B8%8E%E5%85%B6%E4%BD%99%E7%BB%84%E7%9A%84%E6%AF%94%E8%BE%83%E8%BF%9B%E8%A1%8C%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90"><span class="toc-number">9.7.2.</span> <span class="toc-text">Group 0 与其余组的比较进行差异分析</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%B7%A8%E7%B1%BB%E7%BE%A4%E6%AF%94%E8%BE%83%E5%9F%BA%E5%9B%A0"><span class="toc-number">9.7.3.</span> <span class="toc-text">跨类群比较基因</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%A0%B9%E6%8D%AE%E5%B7%B2%E7%9F%A5%E7%9A%84%E7%BB%86%E8%83%9E%E6%A0%87%E8%AE%B0%EF%BC%8C%E6%B3%A8%E9%87%8A%E7%BB%86%E8%83%9E%E7%B1%BB%E5%9E%8B"><span class="toc-number">9.8.</span> <span class="toc-text">根据已知的细胞标记，注释细胞类型</span></a><ol class="toc-child"><li class="toc-item toc-level-4"><a class="toc-link" href="#%E6%B3%95%E4%B8%80%EF%BC%9A"><span class="toc-number">9.8.0.1.</span> <span class="toc-text">法一：</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#%E6%B3%95%E4%BA%8C%EF%BC%9A"><span class="toc-number">9.8.0.2.</span> <span class="toc-text">法二：</span></a></li></ol></li></ol></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%95%B0%E6%8D%AE%E4%BF%9D%E5%AD%98"><span class="toc-number">10.</span> <span class="toc-text">数据保存</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%8F%AF%E8%A7%86%E5%8C%96"><span class="toc-number">11.</span> <span class="toc-text">可视化</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE"><span class="toc-number">11.1.</span> <span class="toc-text">基因表达</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE"><span class="toc-number">11.2.</span> <span class="toc-text">差异基因表达</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E8%81%9A%E7%B1%BB%E6%A0%91"><span class="toc-number">11.3.</span> <span class="toc-text">聚类树</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%9B%B8%E5%85%B3%E6%80%A7"><span class="toc-number">11.4.</span> <span class="toc-text">相关性</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%AE%9E%E6%93%8D"><span class="toc-number">12.</span> <span class="toc-text">实操</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#umap"><span class="toc-number">13.</span> <span class="toc-text">umap</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%89%B9%E6%AC%A1%E5%A4%84%E7%90%86"><span class="toc-number">14.</span> <span class="toc-text">批次处理</span></a></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>最新文章</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2024/07/16/Hash/" title="Hash"><img src="/img/jisuanji.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Hash"/></a><div class="content"><a class="title" href="/2024/07/16/Hash/" title="Hash">Hash</a><time datetime="2024-07-16T03:54:38.000Z" title="发表于 2024-07-16 11:54:38">2024-07-16</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2024/07/10/List/" title="List"><img src="/img/jisuanji.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="List"/></a><div class="content"><a class="title" href="/2024/07/10/List/" title="List">List</a><time datetime="2024-07-10T14:09:41.000Z" title="发表于 2024-07-10 22:09:41">2024-07-10</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2024/06/30/Array/" title="Array"><img src="/img/jisuanji.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Array"/></a><div class="content"><a class="title" href="/2024/06/30/Array/" title="Array">Array</a><time datetime="2024-06-30T11:44:37.000Z" title="发表于 2024-06-30 19:44:37">2024-06-30</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2024/06/02/single-cell/" title="单细胞"><img src="/img/shengwu.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="单细胞"/></a><div class="content"><a class="title" href="/2024/06/02/single-cell/" title="单细胞">单细胞</a><time datetime="2024-06-02T08:10:48.000Z" title="发表于 2024-06-02 16:10:48">2024-06-02</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2024/05/31/hello-world/" title="Hello World">Hello World</a><time datetime="2024-05-31T06:14:30.586Z" title="发表于 2024-05-31 14:14:30">2024-05-31</time></div></div></div></div></div></div></main><footer id="footer" style="background-image: url('/img/shengwu.jpg')"><div id="footer-wrap"><div class="copyright">&copy;2020 - 2024 By Lei Lei</div><div class="framework-info"><span>框架 </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>主题 </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div><div class="footer_custom_text"><footer class="py-5 bg-light"><div class="container text-center"><a href="https://glking123.github.io/" target="_blank"><img src="/img/animal.png" class="d-block mb-4 mx-auto" width="80px"></a><p class="text-secondary col-md-9 mx-auto">傲不可长，欲不可从，志不可满，乐不可极。</p><hr class="my-2"><div class="mb-1"><a href="/" class="text-decoration-none text-secondary" style="margin-right:20px;">主页</a><a href="/archives/" class="text-decoration-none text-secondary mx-3" style="margin-right:20px;">时间轴</a><a href="/about/" class="text-decoration-none text-secondary">关于我</a></p></div></footer></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="阅读模式"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="浅色和深色模式转换"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="单栏和双栏切换"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside-config" type="button" title="设置"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="目录"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="回到顶部"><span class="scroll-percent"></span><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js?v=4.13.0"></script><script src="/js/main.js?v=4.13.0"></script><script src="https://cdn.jsdelivr.net/npm/@fancyapps/ui@5.0.33/dist/fancybox/fancybox.umd.min.js"></script><script src="https://cdn.jsdelivr.net/npm/node-snackbar@0.1.16/dist/snackbar.min.js"></script><div class="js-pjax"></div><script id="canvas_nest" defer="defer" color="0,0,255" opacity="0.6" zIndex="-1" count="88" mobile="false" src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1.1.3/dist/canvas-nest.min.js"></script><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1.1.3/dist/activate-power-mode.min.js"></script><script>POWERMODE.colorful = true;
POWERMODE.shake = false;
POWERMODE.mobile = false;
document.body.addEventListener('input', POWERMODE);
</script><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>